{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "cf9b8624-4cc8-42fa-aaee-b17df33e7438",
   "metadata": {},
   "outputs": [],
   "source": [
    "from pathlib import Path\n",
    "import cfgrib\n",
    "import matplotlib.pyplot as plt\n",
    "import numpy as np\n",
    "import pandas as pd\n",
    "import re\n",
    "import datetime\n",
    "import xarray as xr\n",
    "from typing import Union"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "04452225-cb01-4988-b870-9907f96447fc",
   "metadata": {},
   "source": [
    "This notebook is research for [GitHub issue #344: Convert NWP grib files to Zarr intermediate](https://github.com/openclimatefix/nowcasting_dataset/issues/344).\n",
    "\n",
    "Useful links:\n",
    "\n",
    "* [Met Office's data docs](http://cedadocs.ceda.ac.uk/1334/1/uk_model_data_sheet_lores1.pdf)\n",
    "\n",
    "Done:\n",
    "\n",
    "* Merge Wholesale1 and 2 (2 includes dswrf, lcc, mcc, and hcc)\n",
    "* Remove dimensions we don't care about (e.g. select temperature at 1 meter, not 0 meters)\n",
    "* Reshape images from 1D to 2D.\n",
    "* Do we really need to convert datetimes to unix epochs before appending to zarr?  If no, submit a bug report.\n",
    "* eccodes takes ages to load multiple datasets from each grib.  Maybe it'd be faster to pre-load each grib into ramdisk?  UPDATE: Nope, it's not faster!  What does give a big speedup, though, is using an idx file!\n",
    "* Reshaping is pretty slow.  Maybe go back to using `np.reshape`?\n",
    "* Experiment with when it might be fastest to load data into memory.\n",
    "\n",
    "Some outstanding questions / Todo items\n",
    "\n",
    "* Remove `wholesale_file_number`.  And then use a `pd.Series` instead of a DF.\n",
    "* Zarr chunk size and compression.\n",
    "* Do we need to combine all the DataArrays into a single DataArray (with \"variable\" being a dimension?).  The upside is that then a single Zarr chunk can include multiple variables.  The downside is that we lose the metadata (but that's not a huge problem, maybe?)\n",
    "* Separately log \"bad files\" to a CSV file?\n",
    "* Restart from last `time` in existing Zarr.\n",
    "* Convert to float16?\n",
    "* Check for NaNs.  cdcb has NaNs.\n",
    "* Parallelise\n",
    "* Convert to script\n",
    "* Use click to set source and target directories\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "89fc49d7-2806-4277-9019-c33d4e6a3963",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Define geographical domain for UKV.\n",
    "# Taken from page 4 of http://cedadocs.ceda.ac.uk/1334/1/uk_model_data_sheet_lores1.pdf\n",
    "# To quote the PDF:\n",
    "# \"The United Kingdom domain is a 1,096km x 1,408km ~2km resolution grid.\n",
    "# The OS National Grid corners of the domain are:\"\n",
    "\n",
    "DY_METERS = DX_METERS = 2_000\n",
    "NORTH = 1223_000\n",
    "SOUTH = -185_000\n",
    "WEST = -239_000\n",
    "EAST = 857_000\n",
    "\n",
    "# Note that the UKV NWPs y is top-to-bottom\n",
    "NORTHING = np.arange(start=NORTH, stop=SOUTH, step=-DY_METERS, dtype=np.int32)\n",
    "EASTING = np.arange(start=WEST, stop=EAST, step=DX_METERS, dtype=np.int32)\n",
    "\n",
    "NUM_ROWS = len(NORTHING)\n",
    "NUM_COLS = len(EASTING)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "c703f9a0-12a9-4a40-b64e-8958efdd267f",
   "metadata": {},
   "outputs": [],
   "source": [
    "# NWP_PATH = Path(\"/home/jack/Data/NWP/01\")\n",
    "# NWP_PATH = Path(\"/media/jack/128GB_USB/NWPs\")  # Must not include trailing slash!!!\n",
    "NWP_PATH = Path(\"/mnt/storage_b/data/ocf/solar_pv_nowcasting/nowcasting_dataset_pipeline/NWP/UK_Met_Office/UKV/native\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "26672e96-73da-43a9-af05-bebb160ad46f",
   "metadata": {},
   "outputs": [],
   "source": [
    "assert NWP_PATH.exists()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "2be8d35b-0212-4513-afd7-b070a41692ea",
   "metadata": {},
   "outputs": [],
   "source": [
    "print(\"Getting list of all filenames...\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "0c8d33a5-5300-4393-b9f7-4842e0ea0364",
   "metadata": {},
   "outputs": [],
   "source": [
    "filenames = list(NWP_PATH.glob(\"*/*/*/*Wholesale[12].grib\"))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "4ee7b391-4e8b-4584-af3c-f42d68cd69e2",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "57"
      ]
     },
     "execution_count": 6,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "print(\"Got\", len(filenames), \"filenames\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "2b4c0761-d45e-496b-8c11-da446daa410f",
   "metadata": {},
   "outputs": [],
   "source": [
    "def grib_filename_to_datetime(full_filename: Path) -> datetime.datetime:\n",
    "    \"\"\"Parse the grib filename and return the datetime encoded in the filename.\n",
    "    \n",
    "    Returns a datetime.\n",
    "      For example, if the filename is '202101010000_u1096_ng_umqv_Wholesale1.grib',\n",
    "      then the returned datetime will be datetime(year=2021, month=1, day=1, hour=0, minute=0).\n",
    "      \n",
    "    Raises RuntimeError if the filename does not match the expected regex pattern.\n",
    "    \"\"\"\n",
    "    # Get the base_filename, which will be of the form '202101010000_u1096_ng_umqv_Wholesale1.grib'\n",
    "    base_filename = full_filename.name\n",
    "\n",
    "    # Use regex to match the year, month, day, hour, minute and wholesale_number.  i.e., group the filename like this:\n",
    "    # '(2021)(01)(01)(00)(00)_u1096_ng_umqv_Wholesale(1)'.\n",
    "    # A quick guide to the relevant regex operators:\n",
    "    #   ^ matches the beginning of the string.\n",
    "    #   () defines a group.\n",
    "    #   (?P<name>...) names the group.  We can access the group with `regex_match.groupdict()[name]`.\n",
    "    #   \\d matches a single digit.\n",
    "    #   {n} matches the preceding item n times.\n",
    "    #   . matches any character.\n",
    "    #   $ matches the end of the string.\n",
    "    regex_pattern_string = (\n",
    "        '^'  # Match the beginning of the string.\n",
    "        '(?P<year>\\d{4})'  # Match the year.\n",
    "        '(?P<month>\\d{2})'  # Match the month.\n",
    "        '(?P<day>\\d{2})'  # Match the day.\n",
    "        '(?P<hour>\\d{2})'  # Match the hour.\n",
    "        '(?P<minute>\\d{2})'  # Match the minute.\n",
    "        '_u1096_ng_umqv_Wholesale\\d\\.grib$'  # Match the end of the string (escape the fullstop).\n",
    "    )\n",
    "    regex_pattern = re.compile(regex_pattern_string)\n",
    "    regex_match = regex_pattern.match(base_filename)\n",
    "    if regex_match is None:\n",
    "        raise RuntimeError(f\"Filename '{full_filename}' does not conform to expected regex pattern '{regex_pattern_string}'!\")\n",
    "\n",
    "    # Convert strings to ints:\n",
    "    regex_groups = {key: int(value) for key, value in regex_match.groupdict().items()}\n",
    "\n",
    "    return datetime.datetime(**regex_groups)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "id": "775e59b0-4d52-4700-8ca6-e92c5e7db02c",
   "metadata": {},
   "outputs": [],
   "source": [
    "def decode_and_group_grib_filenames(filenames: list[Path]) -> pd.Series:\n",
    "    \"\"\"Returns a pd.Series where the index is the datetime of the NWP init time.\n",
    "    \n",
    "    And the values are the full_filename of each grib file.\n",
    "    \"\"\"\n",
    "    n_filenames = len(filenames)\n",
    "    nwp_init_datetimes = np.full(shape=n_filenames, fill_value=np.NaN, dtype='datetime64[ns]')\n",
    "    for i, filename in enumerate(filenames):\n",
    "        nwp_init_datetimes[i] = grib_filename_to_datetime(filename)\n",
    "    \n",
    "    # Swap index and values\n",
    "    map_datetime_to_filename = pd.Series(filenames, index=nwp_init_datetimes)\n",
    "    \n",
    "    return map_datetime_to_filename.sort_index()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "id": "34867f22-9980-44f8-8e46-25ae79f2eba6",
   "metadata": {},
   "outputs": [],
   "source": [
    "def load_grib_file(full_filename: Union[Path, str], verbose: bool=False) -> xr.Dataset:\n",
    "    \"\"\"Merges and loads all contiguous xr.Datasets from the grib file.\n",
    "    \n",
    "    Removes unnecessary variables.  Picks heightAboveGround=1meter for temperature.\n",
    "    \n",
    "    Returns an xr.Dataset which has been loaded from disk.  Loading from disk at this point\n",
    "    takes about 2 seconds for a 250 MB grib file, but speeds up reshape_1d_to_2d\n",
    "    from about 7 seconds to 0.5 seconds :)\n",
    "    \n",
    "    Args:\n",
    "      full_filename:  The full filename (including the path) of a single grib file.\n",
    "      verbose:  If True then print out some useful debugging information.\n",
    "    \"\"\"\n",
    "    # The grib files are \"heterogeneous\", so we use cfgrib.open_datasets\n",
    "    # to return a list of contiguous xr.Datasets.\n",
    "    # See https://github.com/ecmwf/cfgrib#automatic-filtering\n",
    "    datasets_from_grib = cfgrib.open_datasets(full_filename)\n",
    "    n_datasets = len(datasets_from_grib)\n",
    "\n",
    "    # Get each dataset into the right shape for merging:\n",
    "    # Loop round the datasets using an index (instead of `for ds in datasets_from_grib`)\n",
    "    # because we will be modifying each dataset:\n",
    "    for i in range(n_datasets):\n",
    "        ds = datasets_from_grib[i]\n",
    "\n",
    "        if verbose:\n",
    "            print(\"\\nDataset\", i, \"before processing:\\n\", ds, \"\\n\")\n",
    "        \n",
    "        # For temperature, we want the temperature at 1 meter above ground,\n",
    "        # not at 0 meters above ground.  The early NWPs (definitely in the 2016-03-22 NWPs),\n",
    "        # heightAboveGround only has 1 entry (\"1\" meter above ground) and `heightAboveGround` isn't set as a dimension for `t`.\n",
    "        # In later NWPs, 'heightAboveGround' has 2 values (0, 1) is a dimension for `t`.\n",
    "        if 't' in ds and 'heightAboveGround' in ds['t'].dims:\n",
    "            ds = ds.sel(heightAboveGround=1)\n",
    "\n",
    "        # Delete unnecessary variables.\n",
    "        vars_to_delete = [\n",
    "            'unknown', 'valid_time', 'heightAboveGround', 'heightAboveGroundLayer',\n",
    "            'atmosphere', 'cloudBase', 'surface', 'meanSea', 'level']\n",
    "        for var_name in vars_to_delete:\n",
    "            try:\n",
    "                del ds[var_name]\n",
    "            except KeyError as e:\n",
    "                if verbose:\n",
    "                    print('var name not in dataset:', e)\n",
    "            else:\n",
    "                if verbose:\n",
    "                    print('Deleted', var_name)\n",
    "\n",
    "        if verbose:\n",
    "            print(\"\\nDataset\", i, \"after processing:\\n\", ds, \"\\n\")\n",
    "            print(\"**************************************************\")\n",
    "            \n",
    "        datasets_from_grib[i] = ds\n",
    "        del ds\n",
    "    \n",
    "    merged_ds = xr.merge(datasets_from_grib)\n",
    "    del datasets_from_grib  # Save memory\n",
    "    return merged_ds.load()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "id": "bd66015d-b4db-4e01-a82e-72726bbc6a71",
   "metadata": {},
   "outputs": [],
   "source": [
    "def reshape_1d_to_2d(dataset: xr.Dataset) -> xr.Dataset:\n",
    "    \"\"\"Convert 1D into 2D array.\n",
    "    \n",
    "    For each `step`, the pixel values in the grib files represent a 2D image.  But, in the grib,\n",
    "    the values are in a flat 1D array (indexed by the `values` dimension).  The ordering of the pixels are row_0, row_1, row_2, etc.\n",
    "\n",
    "    We reshape every data variable at once using this trick.\n",
    "    \"\"\"\n",
    "    # Adapted from https://stackoverflow.com/a/62667154\n",
    "    \n",
    "    # Don't reshape yet.  Instead just create new coordinates,\n",
    "    # which give the `x` and `y` position of each position in the `values` dimension:\n",
    "    dataset = dataset.assign_coords(\n",
    "        {\n",
    "            'x': ('values', np.tile(EASTING, reps=NUM_ROWS)), \n",
    "            'y': ('values', np.repeat(NORTHING, repeats=NUM_COLS))\n",
    "        })\n",
    "    \n",
    "    # Now set \"values\" to be a MultiIndex, indexed by `y` and `x`:\n",
    "    dataset = dataset.set_index(values=(\"y\", \"x\"))\n",
    "    \n",
    "    # Now unstack.  This gets rid of the `values` dimension and indexes\n",
    "    # the data variables using `y` and `x`.\n",
    "    return dataset.unstack(\"values\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "id": "e42183be-67de-4ced-8a72-4ec61b3b8999",
   "metadata": {},
   "outputs": [],
   "source": [
    "def dataset_has_variables(dataset: xr.Dataset) -> bool:\n",
    "    \"\"\"Return True if `dataset` has at least one variable.\"\"\"\n",
    "    return len(dataset.variables) > 0"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 26,
   "id": "52b89ce1-7d2d-4bc0-b7f5-462ae40eb4f5",
   "metadata": {},
   "outputs": [],
   "source": [
    "def append_to_zarr(dataset: xr.Dataset, zarr_path: Union[str, Path]):\n",
    "    print(\"Writing to disk...\")\n",
    "    zarr_path = Path(zarr_path)\n",
    "    if zarr_path.exists():\n",
    "        to_zarr_kwargs = dict(\n",
    "            append_dim = \"time\",\n",
    "        )\n",
    "    else:\n",
    "        to_zarr_kwargs = dict(\n",
    "            # Need to manually set the time units otherwise xarray defaults to using\n",
    "            # units of *days* (and hence cannot represent sub-day temporal resolution), which corrupts\n",
    "            # the `time` values when we appending to Zarr.  See:\n",
    "            # https://github.com/pydata/xarray/issues/5969 and\n",
    "            # http://xarray.pydata.org/en/stable/user-guide/io.html#time-units\n",
    "            encoding={\n",
    "                'time': {\n",
    "                    'units': 'nanoseconds since 1970-01-01'\n",
    "                },\n",
    "            },\n",
    "        )\n",
    "\n",
    "    dataset.to_zarr(zarr_path, **to_zarr_kwargs)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 27,
   "id": "6aff43d0-3e8b-4491-b781-a53f2da113eb",
   "metadata": {},
   "outputs": [],
   "source": [
    "def load_grib_for_single_nwp_init_time(full_filenames: list[Path]) -> xr.Dataset:\n",
    "    datasets_for_nwp_init_datetime = []\n",
    "    for full_filename in full_filenames:\n",
    "        print('Loading', full_filename)\n",
    "        try:\n",
    "            dataset_for_filename = load_grib_file(full_filename)\n",
    "        except EOFError as e:\n",
    "            print(e, f\"Filesize = {full_filename.stat().st_size:,d} bytes\")\n",
    "            # If any of the files associated with this nwp_init_datetime is broken then\n",
    "            # skip all, because we don't want incomplete data for an init_datetime.\n",
    "            datasets_for_nwp_init_datetime = []\n",
    "            break\n",
    "        else:\n",
    "            if dataset_has_variables(dataset_for_filename):\n",
    "                datasets_for_nwp_init_datetime.append(dataset_for_filename)\n",
    "    if len(datasets_for_nwp_init_datetime) == 0:\n",
    "        return\n",
    "    print(\"Merging...\")\n",
    "    dataset_for_nwp_init_datetime = xr.merge(datasets_for_nwp_init_datetime)\n",
    "    del datasets_for_nwp_init_datetime\n",
    "    print(\"Reshaping...\")\n",
    "    dataset_for_nwp_init_datetime = reshape_1d_to_2d(dataset_for_nwp_init_datetime)\n",
    "    return dataset_for_nwp_init_datetime.expand_dims(\"time\", axis=0)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 28,
   "id": "a39717ed-e5ed-4b79-8375-47f0315fb24f",
   "metadata": {},
   "outputs": [],
   "source": [
    "map_datetime_to_grib_filename = decode_and_group_grib_filenames(filenames)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 33,
   "id": "896e2e74-23a9-45fb-9544-a96e6364c2c1",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "2016-03-22 18:00:00    /media/jack/128GB_USB/NWPs/2016/03/22/20160322...\n",
       "2016-03-22 21:00:00    /media/jack/128GB_USB/NWPs/2016/03/22/20160322...\n",
       "2016-03-22 21:00:00    /media/jack/128GB_USB/NWPs/2016/03/22/20160322...\n",
       "2017-01-06 00:00:00    /media/jack/128GB_USB/NWPs/2017/01/06/20170106...\n",
       "2017-01-06 00:00:00    /media/jack/128GB_USB/NWPs/2017/01/06/20170106...\n",
       "dtype: object"
      ]
     },
     "execution_count": 33,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "map_datetime_to_grib_filename.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 47,
   "id": "362eb79c-3912-4517-9e43-7124270d03a4",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "ecCodes provides no latitudes/longitudes for gridType='transverse_mercator'\n",
      "ecCodes provides no latitudes/longitudes for gridType='transverse_mercator'\n",
      "ecCodes provides no latitudes/longitudes for gridType='transverse_mercator'\n",
      "ecCodes provides no latitudes/longitudes for gridType='transverse_mercator'\n",
      "ecCodes provides no latitudes/longitudes for gridType='transverse_mercator'\n",
      "ecCodes provides no latitudes/longitudes for gridType='transverse_mercator'\n",
      "ecCodes provides no latitudes/longitudes for gridType='transverse_mercator'\n",
      "ecCodes provides no latitudes/longitudes for gridType='transverse_mercator'\n",
      "ecCodes provides no latitudes/longitudes for gridType='transverse_mercator'\n",
      "ecCodes provides no latitudes/longitudes for gridType='transverse_mercator'\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "1 filenames found for 2016-03-22 18:00:00 .  Expected 2. Skipping!\n",
      "Loading /media/jack/128GB_USB/NWPs/2016/03/22/201603222100_u1096_ng_umqv_Wholesale1.grib\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "ecCodes provides no latitudes/longitudes for gridType='transverse_mercator'\n",
      "ecCodes provides no latitudes/longitudes for gridType='transverse_mercator'\n",
      "ecCodes provides no latitudes/longitudes for gridType='transverse_mercator'\n",
      "ecCodes provides no latitudes/longitudes for gridType='transverse_mercator'\n",
      "ecCodes provides no latitudes/longitudes for gridType='transverse_mercator'\n",
      "ecCodes provides no latitudes/longitudes for gridType='transverse_mercator'\n",
      "ecCodes provides no latitudes/longitudes for gridType='transverse_mercator'\n",
      "ecCodes provides no latitudes/longitudes for gridType='transverse_mercator'\n",
      "ecCodes provides no latitudes/longitudes for gridType='transverse_mercator'\n",
      "ecCodes provides no latitudes/longitudes for gridType='transverse_mercator'\n",
      "ecCodes provides no latitudes/longitudes for gridType='transverse_mercator'\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Loading /media/jack/128GB_USB/NWPs/2016/03/22/201603222100_u1096_ng_umqv_Wholesale2.grib\n",
      "Merging...\n",
      "Reshaping...\n",
      "Saving to Zarr...\n",
      "Writing to disk...\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Can't create file '/media/jack/128GB_USB/NWPs/2017/01/06/201701060000_u1096_ng_umqv_Wholesale1.grib.923a8.idx'\n",
      "Traceback (most recent call last):\n",
      "  File \"/home/jack/miniconda3/envs/nowcasting_dataset/lib/python3.9/site-packages/cfgrib/messages.py\", line 239, in __iter__\n",
      "    yield self.message_from_file(file, errors=self.errors)\n",
      "  File \"/home/jack/miniconda3/envs/nowcasting_dataset/lib/python3.9/site-packages/cfgrib/messages.py\", line 255, in message_from_file\n",
      "    return self.message_class.from_file(file, offset, **kwargs)\n",
      "  File \"/home/jack/miniconda3/envs/nowcasting_dataset/lib/python3.9/site-packages/cfgrib/messages.py\", line 102, in from_file\n",
      "    raise EOFError(\"End of file: %r\" % file)\n",
      "EOFError: End of file: <_io.BufferedReader name='/media/jack/128GB_USB/NWPs/2017/01/06/201701060000_u1096_ng_umqv_Wholesale1.grib'>\n",
      "\n",
      "During handling of the above exception, another exception occurred:\n",
      "\n",
      "Traceback (most recent call last):\n",
      "  File \"/home/jack/miniconda3/envs/nowcasting_dataset/lib/python3.9/site-packages/cfgrib/messages.py\", line 349, in from_indexpath_or_filestream\n",
      "    self = cls.from_filestream(filestream, index_keys)\n",
      "  File \"/home/jack/miniconda3/envs/nowcasting_dataset/lib/python3.9/site-packages/cfgrib/messages.py\", line 298, in from_filestream\n",
      "    for message in filestream:\n",
      "  File \"/home/jack/miniconda3/envs/nowcasting_dataset/lib/python3.9/site-packages/cfgrib/messages.py\", line 243, in __iter__\n",
      "    raise EOFError(\"No valid message found in file: %r\" % self.path)\n",
      "EOFError: No valid message found in file: '/media/jack/128GB_USB/NWPs/2017/01/06/201701060000_u1096_ng_umqv_Wholesale1.grib'\n",
      "Can't read index file '/media/jack/128GB_USB/NWPs/2017/01/06/201701060000_u1096_ng_umqv_Wholesale1.grib.923a8.idx'\n",
      "Traceback (most recent call last):\n",
      "  File \"/home/jack/miniconda3/envs/nowcasting_dataset/lib/python3.9/site-packages/cfgrib/messages.py\", line 358, in from_indexpath_or_filestream\n",
      "    index_mtime = os.path.getmtime(indexpath)\n",
      "  File \"/home/jack/miniconda3/envs/nowcasting_dataset/lib/python3.9/genericpath.py\", line 55, in getmtime\n",
      "    return os.stat(filename).st_mtime\n",
      "FileNotFoundError: [Errno 2] No such file or directory: '/media/jack/128GB_USB/NWPs/2017/01/06/201701060000_u1096_ng_umqv_Wholesale1.grib.923a8.idx'\n",
      "Can't create file '/media/jack/128GB_USB/NWPs/2017/01/06/201701060300_u1096_ng_umqv_Wholesale1.grib.923a8.idx'\n",
      "Traceback (most recent call last):\n",
      "  File \"/home/jack/miniconda3/envs/nowcasting_dataset/lib/python3.9/site-packages/cfgrib/messages.py\", line 239, in __iter__\n",
      "    yield self.message_from_file(file, errors=self.errors)\n",
      "  File \"/home/jack/miniconda3/envs/nowcasting_dataset/lib/python3.9/site-packages/cfgrib/messages.py\", line 255, in message_from_file\n",
      "    return self.message_class.from_file(file, offset, **kwargs)\n",
      "  File \"/home/jack/miniconda3/envs/nowcasting_dataset/lib/python3.9/site-packages/cfgrib/messages.py\", line 102, in from_file\n",
      "    raise EOFError(\"End of file: %r\" % file)\n",
      "EOFError: End of file: <_io.BufferedReader name='/media/jack/128GB_USB/NWPs/2017/01/06/201701060300_u1096_ng_umqv_Wholesale1.grib'>\n",
      "\n",
      "During handling of the above exception, another exception occurred:\n",
      "\n",
      "Traceback (most recent call last):\n",
      "  File \"/home/jack/miniconda3/envs/nowcasting_dataset/lib/python3.9/site-packages/cfgrib/messages.py\", line 349, in from_indexpath_or_filestream\n",
      "    self = cls.from_filestream(filestream, index_keys)\n",
      "  File \"/home/jack/miniconda3/envs/nowcasting_dataset/lib/python3.9/site-packages/cfgrib/messages.py\", line 298, in from_filestream\n",
      "    for message in filestream:\n",
      "  File \"/home/jack/miniconda3/envs/nowcasting_dataset/lib/python3.9/site-packages/cfgrib/messages.py\", line 243, in __iter__\n",
      "    raise EOFError(\"No valid message found in file: %r\" % self.path)\n",
      "EOFError: No valid message found in file: '/media/jack/128GB_USB/NWPs/2017/01/06/201701060300_u1096_ng_umqv_Wholesale1.grib'\n",
      "Can't read index file '/media/jack/128GB_USB/NWPs/2017/01/06/201701060300_u1096_ng_umqv_Wholesale1.grib.923a8.idx'\n",
      "Traceback (most recent call last):\n",
      "  File \"/home/jack/miniconda3/envs/nowcasting_dataset/lib/python3.9/site-packages/cfgrib/messages.py\", line 358, in from_indexpath_or_filestream\n",
      "    index_mtime = os.path.getmtime(indexpath)\n",
      "  File \"/home/jack/miniconda3/envs/nowcasting_dataset/lib/python3.9/genericpath.py\", line 55, in getmtime\n",
      "    return os.stat(filename).st_mtime\n",
      "FileNotFoundError: [Errno 2] No such file or directory: '/media/jack/128GB_USB/NWPs/2017/01/06/201701060300_u1096_ng_umqv_Wholesale1.grib.923a8.idx'\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Loading /media/jack/128GB_USB/NWPs/2017/01/06/201701060000_u1096_ng_umqv_Wholesale1.grib\n",
      "No valid message found in file: '/media/jack/128GB_USB/NWPs/2017/01/06/201701060000_u1096_ng_umqv_Wholesale1.grib' Filesize = 0 bytes\n",
      "Loading /media/jack/128GB_USB/NWPs/2017/01/06/201701060300_u1096_ng_umqv_Wholesale1.grib\n",
      "No valid message found in file: '/media/jack/128GB_USB/NWPs/2017/01/06/201701060300_u1096_ng_umqv_Wholesale1.grib' Filesize = 0 bytes\n",
      "Loading /media/jack/128GB_USB/NWPs/2017/01/06/201701060600_u1096_ng_umqv_Wholesale1.grib\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "ecCodes provides no latitudes/longitudes for gridType='transverse_mercator'\n",
      "ecCodes provides no latitudes/longitudes for gridType='transverse_mercator'\n",
      "ecCodes provides no latitudes/longitudes for gridType='transverse_mercator'\n",
      "ecCodes provides no latitudes/longitudes for gridType='transverse_mercator'\n",
      "ecCodes provides no latitudes/longitudes for gridType='transverse_mercator'\n",
      "ecCodes provides no latitudes/longitudes for gridType='transverse_mercator'\n",
      "ecCodes provides no latitudes/longitudes for gridType='transverse_mercator'\n",
      "ecCodes provides no latitudes/longitudes for gridType='transverse_mercator'\n",
      "ecCodes provides no latitudes/longitudes for gridType='transverse_mercator'\n",
      "ecCodes provides no latitudes/longitudes for gridType='transverse_mercator'\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Loading /media/jack/128GB_USB/NWPs/2017/01/06/201701060600_u1096_ng_umqv_Wholesale2.grib\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "ecCodes provides no latitudes/longitudes for gridType='transverse_mercator'\n",
      "ecCodes provides no latitudes/longitudes for gridType='transverse_mercator'\n",
      "ecCodes provides no latitudes/longitudes for gridType='transverse_mercator'\n",
      "ecCodes provides no latitudes/longitudes for gridType='transverse_mercator'\n",
      "ecCodes provides no latitudes/longitudes for gridType='transverse_mercator'\n",
      "ecCodes provides no latitudes/longitudes for gridType='transverse_mercator'\n",
      "ecCodes provides no latitudes/longitudes for gridType='transverse_mercator'\n",
      "ecCodes provides no latitudes/longitudes for gridType='transverse_mercator'\n",
      "ecCodes provides no latitudes/longitudes for gridType='transverse_mercator'\n",
      "ecCodes provides no latitudes/longitudes for gridType='transverse_mercator'\n",
      "ecCodes provides no latitudes/longitudes for gridType='transverse_mercator'\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Merging...\n",
      "Reshaping...\n",
      "Saving to Zarr...\n",
      "Writing to disk...\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Can't create file '/media/jack/128GB_USB/NWPs/2017/01/06/201701060900_u1096_ng_umqv_Wholesale2.grib.923a8.idx'\n",
      "Traceback (most recent call last):\n",
      "  File \"/home/jack/miniconda3/envs/nowcasting_dataset/lib/python3.9/site-packages/cfgrib/messages.py\", line 239, in __iter__\n",
      "    yield self.message_from_file(file, errors=self.errors)\n",
      "  File \"/home/jack/miniconda3/envs/nowcasting_dataset/lib/python3.9/site-packages/cfgrib/messages.py\", line 255, in message_from_file\n",
      "    return self.message_class.from_file(file, offset, **kwargs)\n",
      "  File \"/home/jack/miniconda3/envs/nowcasting_dataset/lib/python3.9/site-packages/cfgrib/messages.py\", line 102, in from_file\n",
      "    raise EOFError(\"End of file: %r\" % file)\n",
      "EOFError: End of file: <_io.BufferedReader name='/media/jack/128GB_USB/NWPs/2017/01/06/201701060900_u1096_ng_umqv_Wholesale2.grib'>\n",
      "\n",
      "During handling of the above exception, another exception occurred:\n",
      "\n",
      "Traceback (most recent call last):\n",
      "  File \"/home/jack/miniconda3/envs/nowcasting_dataset/lib/python3.9/site-packages/cfgrib/messages.py\", line 349, in from_indexpath_or_filestream\n",
      "    self = cls.from_filestream(filestream, index_keys)\n",
      "  File \"/home/jack/miniconda3/envs/nowcasting_dataset/lib/python3.9/site-packages/cfgrib/messages.py\", line 298, in from_filestream\n",
      "    for message in filestream:\n",
      "  File \"/home/jack/miniconda3/envs/nowcasting_dataset/lib/python3.9/site-packages/cfgrib/messages.py\", line 243, in __iter__\n",
      "    raise EOFError(\"No valid message found in file: %r\" % self.path)\n",
      "EOFError: No valid message found in file: '/media/jack/128GB_USB/NWPs/2017/01/06/201701060900_u1096_ng_umqv_Wholesale2.grib'\n",
      "Can't read index file '/media/jack/128GB_USB/NWPs/2017/01/06/201701060900_u1096_ng_umqv_Wholesale2.grib.923a8.idx'\n",
      "Traceback (most recent call last):\n",
      "  File \"/home/jack/miniconda3/envs/nowcasting_dataset/lib/python3.9/site-packages/cfgrib/messages.py\", line 358, in from_indexpath_or_filestream\n",
      "    index_mtime = os.path.getmtime(indexpath)\n",
      "  File \"/home/jack/miniconda3/envs/nowcasting_dataset/lib/python3.9/genericpath.py\", line 55, in getmtime\n",
      "    return os.stat(filename).st_mtime\n",
      "FileNotFoundError: [Errno 2] No such file or directory: '/media/jack/128GB_USB/NWPs/2017/01/06/201701060900_u1096_ng_umqv_Wholesale2.grib.923a8.idx'\n",
      "Can't create file '/media/jack/128GB_USB/NWPs/2017/01/06/201701061200_u1096_ng_umqv_Wholesale1.grib.923a8.idx'\n",
      "Traceback (most recent call last):\n",
      "  File \"/home/jack/miniconda3/envs/nowcasting_dataset/lib/python3.9/site-packages/cfgrib/messages.py\", line 239, in __iter__\n",
      "    yield self.message_from_file(file, errors=self.errors)\n",
      "  File \"/home/jack/miniconda3/envs/nowcasting_dataset/lib/python3.9/site-packages/cfgrib/messages.py\", line 255, in message_from_file\n",
      "    return self.message_class.from_file(file, offset, **kwargs)\n",
      "  File \"/home/jack/miniconda3/envs/nowcasting_dataset/lib/python3.9/site-packages/cfgrib/messages.py\", line 102, in from_file\n",
      "    raise EOFError(\"End of file: %r\" % file)\n",
      "EOFError: End of file: <_io.BufferedReader name='/media/jack/128GB_USB/NWPs/2017/01/06/201701061200_u1096_ng_umqv_Wholesale1.grib'>\n",
      "\n",
      "During handling of the above exception, another exception occurred:\n",
      "\n",
      "Traceback (most recent call last):\n",
      "  File \"/home/jack/miniconda3/envs/nowcasting_dataset/lib/python3.9/site-packages/cfgrib/messages.py\", line 349, in from_indexpath_or_filestream\n",
      "    self = cls.from_filestream(filestream, index_keys)\n",
      "  File \"/home/jack/miniconda3/envs/nowcasting_dataset/lib/python3.9/site-packages/cfgrib/messages.py\", line 298, in from_filestream\n",
      "    for message in filestream:\n",
      "  File \"/home/jack/miniconda3/envs/nowcasting_dataset/lib/python3.9/site-packages/cfgrib/messages.py\", line 243, in __iter__\n",
      "    raise EOFError(\"No valid message found in file: %r\" % self.path)\n",
      "EOFError: No valid message found in file: '/media/jack/128GB_USB/NWPs/2017/01/06/201701061200_u1096_ng_umqv_Wholesale1.grib'\n",
      "Can't read index file '/media/jack/128GB_USB/NWPs/2017/01/06/201701061200_u1096_ng_umqv_Wholesale1.grib.923a8.idx'\n",
      "Traceback (most recent call last):\n",
      "  File \"/home/jack/miniconda3/envs/nowcasting_dataset/lib/python3.9/site-packages/cfgrib/messages.py\", line 358, in from_indexpath_or_filestream\n",
      "    index_mtime = os.path.getmtime(indexpath)\n",
      "  File \"/home/jack/miniconda3/envs/nowcasting_dataset/lib/python3.9/genericpath.py\", line 55, in getmtime\n",
      "    return os.stat(filename).st_mtime\n",
      "FileNotFoundError: [Errno 2] No such file or directory: '/media/jack/128GB_USB/NWPs/2017/01/06/201701061200_u1096_ng_umqv_Wholesale1.grib.923a8.idx'\n",
      "Can't create file '/media/jack/128GB_USB/NWPs/2017/01/06/201701061500_u1096_ng_umqv_Wholesale1.grib.923a8.idx'\n",
      "Traceback (most recent call last):\n",
      "  File \"/home/jack/miniconda3/envs/nowcasting_dataset/lib/python3.9/site-packages/cfgrib/messages.py\", line 239, in __iter__\n",
      "    yield self.message_from_file(file, errors=self.errors)\n",
      "  File \"/home/jack/miniconda3/envs/nowcasting_dataset/lib/python3.9/site-packages/cfgrib/messages.py\", line 255, in message_from_file\n",
      "    return self.message_class.from_file(file, offset, **kwargs)\n",
      "  File \"/home/jack/miniconda3/envs/nowcasting_dataset/lib/python3.9/site-packages/cfgrib/messages.py\", line 102, in from_file\n",
      "    raise EOFError(\"End of file: %r\" % file)\n",
      "EOFError: End of file: <_io.BufferedReader name='/media/jack/128GB_USB/NWPs/2017/01/06/201701061500_u1096_ng_umqv_Wholesale1.grib'>\n",
      "\n",
      "During handling of the above exception, another exception occurred:\n",
      "\n",
      "Traceback (most recent call last):\n",
      "  File \"/home/jack/miniconda3/envs/nowcasting_dataset/lib/python3.9/site-packages/cfgrib/messages.py\", line 349, in from_indexpath_or_filestream\n",
      "    self = cls.from_filestream(filestream, index_keys)\n",
      "  File \"/home/jack/miniconda3/envs/nowcasting_dataset/lib/python3.9/site-packages/cfgrib/messages.py\", line 298, in from_filestream\n",
      "    for message in filestream:\n",
      "  File \"/home/jack/miniconda3/envs/nowcasting_dataset/lib/python3.9/site-packages/cfgrib/messages.py\", line 243, in __iter__\n",
      "    raise EOFError(\"No valid message found in file: %r\" % self.path)\n",
      "EOFError: No valid message found in file: '/media/jack/128GB_USB/NWPs/2017/01/06/201701061500_u1096_ng_umqv_Wholesale1.grib'\n",
      "Can't read index file '/media/jack/128GB_USB/NWPs/2017/01/06/201701061500_u1096_ng_umqv_Wholesale1.grib.923a8.idx'\n",
      "Traceback (most recent call last):\n",
      "  File \"/home/jack/miniconda3/envs/nowcasting_dataset/lib/python3.9/site-packages/cfgrib/messages.py\", line 358, in from_indexpath_or_filestream\n",
      "    index_mtime = os.path.getmtime(indexpath)\n",
      "  File \"/home/jack/miniconda3/envs/nowcasting_dataset/lib/python3.9/genericpath.py\", line 55, in getmtime\n",
      "    return os.stat(filename).st_mtime\n",
      "FileNotFoundError: [Errno 2] No such file or directory: '/media/jack/128GB_USB/NWPs/2017/01/06/201701061500_u1096_ng_umqv_Wholesale1.grib.923a8.idx'\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Loading /media/jack/128GB_USB/NWPs/2017/01/06/201701060900_u1096_ng_umqv_Wholesale2.grib\n",
      "No valid message found in file: '/media/jack/128GB_USB/NWPs/2017/01/06/201701060900_u1096_ng_umqv_Wholesale2.grib' Filesize = 0 bytes\n",
      "Loading /media/jack/128GB_USB/NWPs/2017/01/06/201701061200_u1096_ng_umqv_Wholesale1.grib\n",
      "No valid message found in file: '/media/jack/128GB_USB/NWPs/2017/01/06/201701061200_u1096_ng_umqv_Wholesale1.grib' Filesize = 0 bytes\n",
      "Loading /media/jack/128GB_USB/NWPs/2017/01/06/201701061500_u1096_ng_umqv_Wholesale1.grib\n",
      "No valid message found in file: '/media/jack/128GB_USB/NWPs/2017/01/06/201701061500_u1096_ng_umqv_Wholesale1.grib' Filesize = 0 bytes\n",
      "Loading /media/jack/128GB_USB/NWPs/2017/01/06/201701061800_u1096_ng_umqv_Wholesale2.grib\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "ecCodes provides no latitudes/longitudes for gridType='transverse_mercator'\n",
      "ecCodes provides no latitudes/longitudes for gridType='transverse_mercator'\n",
      "ecCodes provides no latitudes/longitudes for gridType='transverse_mercator'\n",
      "ecCodes provides no latitudes/longitudes for gridType='transverse_mercator'\n",
      "ecCodes provides no latitudes/longitudes for gridType='transverse_mercator'\n",
      "ecCodes provides no latitudes/longitudes for gridType='transverse_mercator'\n",
      "ecCodes provides no latitudes/longitudes for gridType='transverse_mercator'\n",
      "ecCodes provides no latitudes/longitudes for gridType='transverse_mercator'\n",
      "ecCodes provides no latitudes/longitudes for gridType='transverse_mercator'\n",
      "ecCodes provides no latitudes/longitudes for gridType='transverse_mercator'\n",
      "ecCodes provides no latitudes/longitudes for gridType='transverse_mercator'\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Loading /media/jack/128GB_USB/NWPs/2017/01/06/201701061800_u1096_ng_umqv_Wholesale1.grib\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "ecCodes provides no latitudes/longitudes for gridType='transverse_mercator'\n",
      "ecCodes provides no latitudes/longitudes for gridType='transverse_mercator'\n",
      "ecCodes provides no latitudes/longitudes for gridType='transverse_mercator'\n",
      "ecCodes provides no latitudes/longitudes for gridType='transverse_mercator'\n",
      "ecCodes provides no latitudes/longitudes for gridType='transverse_mercator'\n",
      "ecCodes provides no latitudes/longitudes for gridType='transverse_mercator'\n",
      "ecCodes provides no latitudes/longitudes for gridType='transverse_mercator'\n",
      "ecCodes provides no latitudes/longitudes for gridType='transverse_mercator'\n",
      "ecCodes provides no latitudes/longitudes for gridType='transverse_mercator'\n",
      "ecCodes provides no latitudes/longitudes for gridType='transverse_mercator'\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Merging...\n",
      "Reshaping...\n",
      "Saving to Zarr...\n",
      "Writing to disk...\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Can't create file '/media/jack/128GB_USB/NWPs/2017/01/06/201701062100_u1096_ng_umqv_Wholesale1.grib.923a8.idx'\n",
      "Traceback (most recent call last):\n",
      "  File \"/home/jack/miniconda3/envs/nowcasting_dataset/lib/python3.9/site-packages/cfgrib/messages.py\", line 239, in __iter__\n",
      "    yield self.message_from_file(file, errors=self.errors)\n",
      "  File \"/home/jack/miniconda3/envs/nowcasting_dataset/lib/python3.9/site-packages/cfgrib/messages.py\", line 255, in message_from_file\n",
      "    return self.message_class.from_file(file, offset, **kwargs)\n",
      "  File \"/home/jack/miniconda3/envs/nowcasting_dataset/lib/python3.9/site-packages/cfgrib/messages.py\", line 102, in from_file\n",
      "    raise EOFError(\"End of file: %r\" % file)\n",
      "EOFError: End of file: <_io.BufferedReader name='/media/jack/128GB_USB/NWPs/2017/01/06/201701062100_u1096_ng_umqv_Wholesale1.grib'>\n",
      "\n",
      "During handling of the above exception, another exception occurred:\n",
      "\n",
      "Traceback (most recent call last):\n",
      "  File \"/home/jack/miniconda3/envs/nowcasting_dataset/lib/python3.9/site-packages/cfgrib/messages.py\", line 349, in from_indexpath_or_filestream\n",
      "    self = cls.from_filestream(filestream, index_keys)\n",
      "  File \"/home/jack/miniconda3/envs/nowcasting_dataset/lib/python3.9/site-packages/cfgrib/messages.py\", line 298, in from_filestream\n",
      "    for message in filestream:\n",
      "  File \"/home/jack/miniconda3/envs/nowcasting_dataset/lib/python3.9/site-packages/cfgrib/messages.py\", line 243, in __iter__\n",
      "    raise EOFError(\"No valid message found in file: %r\" % self.path)\n",
      "EOFError: No valid message found in file: '/media/jack/128GB_USB/NWPs/2017/01/06/201701062100_u1096_ng_umqv_Wholesale1.grib'\n",
      "Can't read index file '/media/jack/128GB_USB/NWPs/2017/01/06/201701062100_u1096_ng_umqv_Wholesale1.grib.923a8.idx'\n",
      "Traceback (most recent call last):\n",
      "  File \"/home/jack/miniconda3/envs/nowcasting_dataset/lib/python3.9/site-packages/cfgrib/messages.py\", line 358, in from_indexpath_or_filestream\n",
      "    index_mtime = os.path.getmtime(indexpath)\n",
      "  File \"/home/jack/miniconda3/envs/nowcasting_dataset/lib/python3.9/genericpath.py\", line 55, in getmtime\n",
      "    return os.stat(filename).st_mtime\n",
      "FileNotFoundError: [Errno 2] No such file or directory: '/media/jack/128GB_USB/NWPs/2017/01/06/201701062100_u1096_ng_umqv_Wholesale1.grib.923a8.idx'\n",
      "Can't create file '/media/jack/128GB_USB/NWPs/2018/12/31/201812310000_u1096_ng_umqv_Wholesale2.grib.923a8.idx'\n",
      "Traceback (most recent call last):\n",
      "  File \"/home/jack/miniconda3/envs/nowcasting_dataset/lib/python3.9/site-packages/cfgrib/messages.py\", line 239, in __iter__\n",
      "    yield self.message_from_file(file, errors=self.errors)\n",
      "  File \"/home/jack/miniconda3/envs/nowcasting_dataset/lib/python3.9/site-packages/cfgrib/messages.py\", line 255, in message_from_file\n",
      "    return self.message_class.from_file(file, offset, **kwargs)\n",
      "  File \"/home/jack/miniconda3/envs/nowcasting_dataset/lib/python3.9/site-packages/cfgrib/messages.py\", line 102, in from_file\n",
      "    raise EOFError(\"End of file: %r\" % file)\n",
      "EOFError: End of file: <_io.BufferedReader name='/media/jack/128GB_USB/NWPs/2018/12/31/201812310000_u1096_ng_umqv_Wholesale2.grib'>\n",
      "\n",
      "During handling of the above exception, another exception occurred:\n",
      "\n",
      "Traceback (most recent call last):\n",
      "  File \"/home/jack/miniconda3/envs/nowcasting_dataset/lib/python3.9/site-packages/cfgrib/messages.py\", line 349, in from_indexpath_or_filestream\n",
      "    self = cls.from_filestream(filestream, index_keys)\n",
      "  File \"/home/jack/miniconda3/envs/nowcasting_dataset/lib/python3.9/site-packages/cfgrib/messages.py\", line 298, in from_filestream\n",
      "    for message in filestream:\n",
      "  File \"/home/jack/miniconda3/envs/nowcasting_dataset/lib/python3.9/site-packages/cfgrib/messages.py\", line 243, in __iter__\n",
      "    raise EOFError(\"No valid message found in file: %r\" % self.path)\n",
      "EOFError: No valid message found in file: '/media/jack/128GB_USB/NWPs/2018/12/31/201812310000_u1096_ng_umqv_Wholesale2.grib'\n",
      "Can't read index file '/media/jack/128GB_USB/NWPs/2018/12/31/201812310000_u1096_ng_umqv_Wholesale2.grib.923a8.idx'\n",
      "Traceback (most recent call last):\n",
      "  File \"/home/jack/miniconda3/envs/nowcasting_dataset/lib/python3.9/site-packages/cfgrib/messages.py\", line 358, in from_indexpath_or_filestream\n",
      "    index_mtime = os.path.getmtime(indexpath)\n",
      "  File \"/home/jack/miniconda3/envs/nowcasting_dataset/lib/python3.9/genericpath.py\", line 55, in getmtime\n",
      "    return os.stat(filename).st_mtime\n",
      "FileNotFoundError: [Errno 2] No such file or directory: '/media/jack/128GB_USB/NWPs/2018/12/31/201812310000_u1096_ng_umqv_Wholesale2.grib.923a8.idx'\n",
      "Can't create file '/media/jack/128GB_USB/NWPs/2018/12/31/201812310300_u1096_ng_umqv_Wholesale1.grib.923a8.idx'\n",
      "Traceback (most recent call last):\n",
      "  File \"/home/jack/miniconda3/envs/nowcasting_dataset/lib/python3.9/site-packages/cfgrib/messages.py\", line 239, in __iter__\n",
      "    yield self.message_from_file(file, errors=self.errors)\n",
      "  File \"/home/jack/miniconda3/envs/nowcasting_dataset/lib/python3.9/site-packages/cfgrib/messages.py\", line 255, in message_from_file\n",
      "    return self.message_class.from_file(file, offset, **kwargs)\n",
      "  File \"/home/jack/miniconda3/envs/nowcasting_dataset/lib/python3.9/site-packages/cfgrib/messages.py\", line 102, in from_file\n",
      "    raise EOFError(\"End of file: %r\" % file)\n",
      "EOFError: End of file: <_io.BufferedReader name='/media/jack/128GB_USB/NWPs/2018/12/31/201812310300_u1096_ng_umqv_Wholesale1.grib'>\n",
      "\n",
      "During handling of the above exception, another exception occurred:\n",
      "\n",
      "Traceback (most recent call last):\n",
      "  File \"/home/jack/miniconda3/envs/nowcasting_dataset/lib/python3.9/site-packages/cfgrib/messages.py\", line 349, in from_indexpath_or_filestream\n",
      "    self = cls.from_filestream(filestream, index_keys)\n",
      "  File \"/home/jack/miniconda3/envs/nowcasting_dataset/lib/python3.9/site-packages/cfgrib/messages.py\", line 298, in from_filestream\n",
      "    for message in filestream:\n",
      "  File \"/home/jack/miniconda3/envs/nowcasting_dataset/lib/python3.9/site-packages/cfgrib/messages.py\", line 243, in __iter__\n",
      "    raise EOFError(\"No valid message found in file: %r\" % self.path)\n",
      "EOFError: No valid message found in file: '/media/jack/128GB_USB/NWPs/2018/12/31/201812310300_u1096_ng_umqv_Wholesale1.grib'\n",
      "Can't read index file '/media/jack/128GB_USB/NWPs/2018/12/31/201812310300_u1096_ng_umqv_Wholesale1.grib.923a8.idx'\n",
      "Traceback (most recent call last):\n",
      "  File \"/home/jack/miniconda3/envs/nowcasting_dataset/lib/python3.9/site-packages/cfgrib/messages.py\", line 358, in from_indexpath_or_filestream\n",
      "    index_mtime = os.path.getmtime(indexpath)\n",
      "  File \"/home/jack/miniconda3/envs/nowcasting_dataset/lib/python3.9/genericpath.py\", line 55, in getmtime\n",
      "    return os.stat(filename).st_mtime\n",
      "FileNotFoundError: [Errno 2] No such file or directory: '/media/jack/128GB_USB/NWPs/2018/12/31/201812310300_u1096_ng_umqv_Wholesale1.grib.923a8.idx'\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Loading /media/jack/128GB_USB/NWPs/2017/01/06/201701062100_u1096_ng_umqv_Wholesale1.grib\n",
      "No valid message found in file: '/media/jack/128GB_USB/NWPs/2017/01/06/201701062100_u1096_ng_umqv_Wholesale1.grib' Filesize = 0 bytes\n",
      "Loading /media/jack/128GB_USB/NWPs/2018/12/31/201812310000_u1096_ng_umqv_Wholesale2.grib\n",
      "No valid message found in file: '/media/jack/128GB_USB/NWPs/2018/12/31/201812310000_u1096_ng_umqv_Wholesale2.grib' Filesize = 0 bytes\n",
      "Loading /media/jack/128GB_USB/NWPs/2018/12/31/201812310300_u1096_ng_umqv_Wholesale1.grib\n",
      "No valid message found in file: '/media/jack/128GB_USB/NWPs/2018/12/31/201812310300_u1096_ng_umqv_Wholesale1.grib' Filesize = 0 bytes\n",
      "Loading /media/jack/128GB_USB/NWPs/2018/12/31/201812310600_u1096_ng_umqv_Wholesale1.grib\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "ecCodes provides no latitudes/longitudes for gridType='transverse_mercator'\n",
      "ecCodes provides no latitudes/longitudes for gridType='transverse_mercator'\n",
      "ecCodes provides no latitudes/longitudes for gridType='transverse_mercator'\n",
      "ecCodes provides no latitudes/longitudes for gridType='transverse_mercator'\n",
      "ecCodes provides no latitudes/longitudes for gridType='transverse_mercator'\n",
      "ecCodes provides no latitudes/longitudes for gridType='transverse_mercator'\n",
      "ecCodes provides no latitudes/longitudes for gridType='transverse_mercator'\n",
      "ecCodes provides no latitudes/longitudes for gridType='transverse_mercator'\n",
      "ecCodes provides no latitudes/longitudes for gridType='transverse_mercator'\n",
      "ecCodes provides no latitudes/longitudes for gridType='transverse_mercator'\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Loading /media/jack/128GB_USB/NWPs/2018/12/31/201812310600_u1096_ng_umqv_Wholesale2.grib\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "ecCodes provides no latitudes/longitudes for gridType='transverse_mercator'\n",
      "ecCodes provides no latitudes/longitudes for gridType='transverse_mercator'\n",
      "ecCodes provides no latitudes/longitudes for gridType='transverse_mercator'\n",
      "ecCodes provides no latitudes/longitudes for gridType='transverse_mercator'\n",
      "ecCodes provides no latitudes/longitudes for gridType='transverse_mercator'\n",
      "ecCodes provides no latitudes/longitudes for gridType='transverse_mercator'\n",
      "ecCodes provides no latitudes/longitudes for gridType='transverse_mercator'\n",
      "ecCodes provides no latitudes/longitudes for gridType='transverse_mercator'\n",
      "ecCodes provides no latitudes/longitudes for gridType='transverse_mercator'\n",
      "ecCodes provides no latitudes/longitudes for gridType='transverse_mercator'\n",
      "ecCodes provides no latitudes/longitudes for gridType='transverse_mercator'\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Merging...\n",
      "Reshaping...\n",
      "Saving to Zarr...\n",
      "Writing to disk...\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Can't create file '/media/jack/128GB_USB/NWPs/2018/12/31/201812310900_u1096_ng_umqv_Wholesale2.grib.923a8.idx'\n",
      "Traceback (most recent call last):\n",
      "  File \"/home/jack/miniconda3/envs/nowcasting_dataset/lib/python3.9/site-packages/cfgrib/messages.py\", line 239, in __iter__\n",
      "    yield self.message_from_file(file, errors=self.errors)\n",
      "  File \"/home/jack/miniconda3/envs/nowcasting_dataset/lib/python3.9/site-packages/cfgrib/messages.py\", line 255, in message_from_file\n",
      "    return self.message_class.from_file(file, offset, **kwargs)\n",
      "  File \"/home/jack/miniconda3/envs/nowcasting_dataset/lib/python3.9/site-packages/cfgrib/messages.py\", line 102, in from_file\n",
      "    raise EOFError(\"End of file: %r\" % file)\n",
      "EOFError: End of file: <_io.BufferedReader name='/media/jack/128GB_USB/NWPs/2018/12/31/201812310900_u1096_ng_umqv_Wholesale2.grib'>\n",
      "\n",
      "During handling of the above exception, another exception occurred:\n",
      "\n",
      "Traceback (most recent call last):\n",
      "  File \"/home/jack/miniconda3/envs/nowcasting_dataset/lib/python3.9/site-packages/cfgrib/messages.py\", line 349, in from_indexpath_or_filestream\n",
      "    self = cls.from_filestream(filestream, index_keys)\n",
      "  File \"/home/jack/miniconda3/envs/nowcasting_dataset/lib/python3.9/site-packages/cfgrib/messages.py\", line 298, in from_filestream\n",
      "    for message in filestream:\n",
      "  File \"/home/jack/miniconda3/envs/nowcasting_dataset/lib/python3.9/site-packages/cfgrib/messages.py\", line 243, in __iter__\n",
      "    raise EOFError(\"No valid message found in file: %r\" % self.path)\n",
      "EOFError: No valid message found in file: '/media/jack/128GB_USB/NWPs/2018/12/31/201812310900_u1096_ng_umqv_Wholesale2.grib'\n",
      "Can't read index file '/media/jack/128GB_USB/NWPs/2018/12/31/201812310900_u1096_ng_umqv_Wholesale2.grib.923a8.idx'\n",
      "Traceback (most recent call last):\n",
      "  File \"/home/jack/miniconda3/envs/nowcasting_dataset/lib/python3.9/site-packages/cfgrib/messages.py\", line 358, in from_indexpath_or_filestream\n",
      "    index_mtime = os.path.getmtime(indexpath)\n",
      "  File \"/home/jack/miniconda3/envs/nowcasting_dataset/lib/python3.9/genericpath.py\", line 55, in getmtime\n",
      "    return os.stat(filename).st_mtime\n",
      "FileNotFoundError: [Errno 2] No such file or directory: '/media/jack/128GB_USB/NWPs/2018/12/31/201812310900_u1096_ng_umqv_Wholesale2.grib.923a8.idx'\n",
      "Can't create file '/media/jack/128GB_USB/NWPs/2018/12/31/201812311200_u1096_ng_umqv_Wholesale1.grib.923a8.idx'\n",
      "Traceback (most recent call last):\n",
      "  File \"/home/jack/miniconda3/envs/nowcasting_dataset/lib/python3.9/site-packages/cfgrib/messages.py\", line 239, in __iter__\n",
      "    yield self.message_from_file(file, errors=self.errors)\n",
      "  File \"/home/jack/miniconda3/envs/nowcasting_dataset/lib/python3.9/site-packages/cfgrib/messages.py\", line 255, in message_from_file\n",
      "    return self.message_class.from_file(file, offset, **kwargs)\n",
      "  File \"/home/jack/miniconda3/envs/nowcasting_dataset/lib/python3.9/site-packages/cfgrib/messages.py\", line 102, in from_file\n",
      "    raise EOFError(\"End of file: %r\" % file)\n",
      "EOFError: End of file: <_io.BufferedReader name='/media/jack/128GB_USB/NWPs/2018/12/31/201812311200_u1096_ng_umqv_Wholesale1.grib'>\n",
      "\n",
      "During handling of the above exception, another exception occurred:\n",
      "\n",
      "Traceback (most recent call last):\n",
      "  File \"/home/jack/miniconda3/envs/nowcasting_dataset/lib/python3.9/site-packages/cfgrib/messages.py\", line 349, in from_indexpath_or_filestream\n",
      "    self = cls.from_filestream(filestream, index_keys)\n",
      "  File \"/home/jack/miniconda3/envs/nowcasting_dataset/lib/python3.9/site-packages/cfgrib/messages.py\", line 298, in from_filestream\n",
      "    for message in filestream:\n",
      "  File \"/home/jack/miniconda3/envs/nowcasting_dataset/lib/python3.9/site-packages/cfgrib/messages.py\", line 243, in __iter__\n",
      "    raise EOFError(\"No valid message found in file: %r\" % self.path)\n",
      "EOFError: No valid message found in file: '/media/jack/128GB_USB/NWPs/2018/12/31/201812311200_u1096_ng_umqv_Wholesale1.grib'\n",
      "Can't read index file '/media/jack/128GB_USB/NWPs/2018/12/31/201812311200_u1096_ng_umqv_Wholesale1.grib.923a8.idx'\n",
      "Traceback (most recent call last):\n",
      "  File \"/home/jack/miniconda3/envs/nowcasting_dataset/lib/python3.9/site-packages/cfgrib/messages.py\", line 358, in from_indexpath_or_filestream\n",
      "    index_mtime = os.path.getmtime(indexpath)\n",
      "  File \"/home/jack/miniconda3/envs/nowcasting_dataset/lib/python3.9/genericpath.py\", line 55, in getmtime\n",
      "    return os.stat(filename).st_mtime\n",
      "FileNotFoundError: [Errno 2] No such file or directory: '/media/jack/128GB_USB/NWPs/2018/12/31/201812311200_u1096_ng_umqv_Wholesale1.grib.923a8.idx'\n",
      "Can't create file '/media/jack/128GB_USB/NWPs/2018/12/31/201812311500_u1096_ng_umqv_Wholesale2.grib.923a8.idx'\n",
      "Traceback (most recent call last):\n",
      "  File \"/home/jack/miniconda3/envs/nowcasting_dataset/lib/python3.9/site-packages/cfgrib/messages.py\", line 239, in __iter__\n",
      "    yield self.message_from_file(file, errors=self.errors)\n",
      "  File \"/home/jack/miniconda3/envs/nowcasting_dataset/lib/python3.9/site-packages/cfgrib/messages.py\", line 255, in message_from_file\n",
      "    return self.message_class.from_file(file, offset, **kwargs)\n",
      "  File \"/home/jack/miniconda3/envs/nowcasting_dataset/lib/python3.9/site-packages/cfgrib/messages.py\", line 102, in from_file\n",
      "    raise EOFError(\"End of file: %r\" % file)\n",
      "EOFError: End of file: <_io.BufferedReader name='/media/jack/128GB_USB/NWPs/2018/12/31/201812311500_u1096_ng_umqv_Wholesale2.grib'>\n",
      "\n",
      "During handling of the above exception, another exception occurred:\n",
      "\n",
      "Traceback (most recent call last):\n",
      "  File \"/home/jack/miniconda3/envs/nowcasting_dataset/lib/python3.9/site-packages/cfgrib/messages.py\", line 349, in from_indexpath_or_filestream\n",
      "    self = cls.from_filestream(filestream, index_keys)\n",
      "  File \"/home/jack/miniconda3/envs/nowcasting_dataset/lib/python3.9/site-packages/cfgrib/messages.py\", line 298, in from_filestream\n",
      "    for message in filestream:\n",
      "  File \"/home/jack/miniconda3/envs/nowcasting_dataset/lib/python3.9/site-packages/cfgrib/messages.py\", line 243, in __iter__\n",
      "    raise EOFError(\"No valid message found in file: %r\" % self.path)\n",
      "EOFError: No valid message found in file: '/media/jack/128GB_USB/NWPs/2018/12/31/201812311500_u1096_ng_umqv_Wholesale2.grib'\n",
      "Can't read index file '/media/jack/128GB_USB/NWPs/2018/12/31/201812311500_u1096_ng_umqv_Wholesale2.grib.923a8.idx'\n",
      "Traceback (most recent call last):\n",
      "  File \"/home/jack/miniconda3/envs/nowcasting_dataset/lib/python3.9/site-packages/cfgrib/messages.py\", line 358, in from_indexpath_or_filestream\n",
      "    index_mtime = os.path.getmtime(indexpath)\n",
      "  File \"/home/jack/miniconda3/envs/nowcasting_dataset/lib/python3.9/genericpath.py\", line 55, in getmtime\n",
      "    return os.stat(filename).st_mtime\n",
      "FileNotFoundError: [Errno 2] No such file or directory: '/media/jack/128GB_USB/NWPs/2018/12/31/201812311500_u1096_ng_umqv_Wholesale2.grib.923a8.idx'\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Loading /media/jack/128GB_USB/NWPs/2018/12/31/201812310900_u1096_ng_umqv_Wholesale2.grib\n",
      "No valid message found in file: '/media/jack/128GB_USB/NWPs/2018/12/31/201812310900_u1096_ng_umqv_Wholesale2.grib' Filesize = 0 bytes\n",
      "Loading /media/jack/128GB_USB/NWPs/2018/12/31/201812311200_u1096_ng_umqv_Wholesale1.grib\n",
      "No valid message found in file: '/media/jack/128GB_USB/NWPs/2018/12/31/201812311200_u1096_ng_umqv_Wholesale1.grib' Filesize = 0 bytes\n",
      "Loading /media/jack/128GB_USB/NWPs/2018/12/31/201812311500_u1096_ng_umqv_Wholesale2.grib\n",
      "No valid message found in file: '/media/jack/128GB_USB/NWPs/2018/12/31/201812311500_u1096_ng_umqv_Wholesale2.grib' Filesize = 0 bytes\n",
      "Loading /media/jack/128GB_USB/NWPs/2018/12/31/201812311800_u1096_ng_umqv_Wholesale2.grib\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "ecCodes provides no latitudes/longitudes for gridType='transverse_mercator'\n",
      "ecCodes provides no latitudes/longitudes for gridType='transverse_mercator'\n",
      "ecCodes provides no latitudes/longitudes for gridType='transverse_mercator'\n",
      "ecCodes provides no latitudes/longitudes for gridType='transverse_mercator'\n",
      "ecCodes provides no latitudes/longitudes for gridType='transverse_mercator'\n",
      "ecCodes provides no latitudes/longitudes for gridType='transverse_mercator'\n",
      "ecCodes provides no latitudes/longitudes for gridType='transverse_mercator'\n",
      "ecCodes provides no latitudes/longitudes for gridType='transverse_mercator'\n",
      "ecCodes provides no latitudes/longitudes for gridType='transverse_mercator'\n",
      "ecCodes provides no latitudes/longitudes for gridType='transverse_mercator'\n",
      "ecCodes provides no latitudes/longitudes for gridType='transverse_mercator'\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Loading /media/jack/128GB_USB/NWPs/2018/12/31/201812311800_u1096_ng_umqv_Wholesale1.grib\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "ecCodes provides no latitudes/longitudes for gridType='transverse_mercator'\n",
      "ecCodes provides no latitudes/longitudes for gridType='transverse_mercator'\n",
      "ecCodes provides no latitudes/longitudes for gridType='transverse_mercator'\n",
      "ecCodes provides no latitudes/longitudes for gridType='transverse_mercator'\n",
      "ecCodes provides no latitudes/longitudes for gridType='transverse_mercator'\n",
      "ecCodes provides no latitudes/longitudes for gridType='transverse_mercator'\n",
      "ecCodes provides no latitudes/longitudes for gridType='transverse_mercator'\n",
      "ecCodes provides no latitudes/longitudes for gridType='transverse_mercator'\n",
      "ecCodes provides no latitudes/longitudes for gridType='transverse_mercator'\n",
      "ecCodes provides no latitudes/longitudes for gridType='transverse_mercator'\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Merging...\n",
      "Reshaping...\n",
      "Saving to Zarr...\n",
      "Writing to disk...\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Can't create file '/media/jack/128GB_USB/NWPs/2018/12/31/201812312100_u1096_ng_umqv_Wholesale1.grib.923a8.idx'\n",
      "Traceback (most recent call last):\n",
      "  File \"/home/jack/miniconda3/envs/nowcasting_dataset/lib/python3.9/site-packages/cfgrib/messages.py\", line 239, in __iter__\n",
      "    yield self.message_from_file(file, errors=self.errors)\n",
      "  File \"/home/jack/miniconda3/envs/nowcasting_dataset/lib/python3.9/site-packages/cfgrib/messages.py\", line 255, in message_from_file\n",
      "    return self.message_class.from_file(file, offset, **kwargs)\n",
      "  File \"/home/jack/miniconda3/envs/nowcasting_dataset/lib/python3.9/site-packages/cfgrib/messages.py\", line 102, in from_file\n",
      "    raise EOFError(\"End of file: %r\" % file)\n",
      "EOFError: End of file: <_io.BufferedReader name='/media/jack/128GB_USB/NWPs/2018/12/31/201812312100_u1096_ng_umqv_Wholesale1.grib'>\n",
      "\n",
      "During handling of the above exception, another exception occurred:\n",
      "\n",
      "Traceback (most recent call last):\n",
      "  File \"/home/jack/miniconda3/envs/nowcasting_dataset/lib/python3.9/site-packages/cfgrib/messages.py\", line 349, in from_indexpath_or_filestream\n",
      "    self = cls.from_filestream(filestream, index_keys)\n",
      "  File \"/home/jack/miniconda3/envs/nowcasting_dataset/lib/python3.9/site-packages/cfgrib/messages.py\", line 298, in from_filestream\n",
      "    for message in filestream:\n",
      "  File \"/home/jack/miniconda3/envs/nowcasting_dataset/lib/python3.9/site-packages/cfgrib/messages.py\", line 243, in __iter__\n",
      "    raise EOFError(\"No valid message found in file: %r\" % self.path)\n",
      "EOFError: No valid message found in file: '/media/jack/128GB_USB/NWPs/2018/12/31/201812312100_u1096_ng_umqv_Wholesale1.grib'\n",
      "Can't read index file '/media/jack/128GB_USB/NWPs/2018/12/31/201812312100_u1096_ng_umqv_Wholesale1.grib.923a8.idx'\n",
      "Traceback (most recent call last):\n",
      "  File \"/home/jack/miniconda3/envs/nowcasting_dataset/lib/python3.9/site-packages/cfgrib/messages.py\", line 358, in from_indexpath_or_filestream\n",
      "    index_mtime = os.path.getmtime(indexpath)\n",
      "  File \"/home/jack/miniconda3/envs/nowcasting_dataset/lib/python3.9/genericpath.py\", line 55, in getmtime\n",
      "    return os.stat(filename).st_mtime\n",
      "FileNotFoundError: [Errno 2] No such file or directory: '/media/jack/128GB_USB/NWPs/2018/12/31/201812312100_u1096_ng_umqv_Wholesale1.grib.923a8.idx'\n",
      "Can't create file '/media/jack/128GB_USB/NWPs/2019/06/15/201906150000_u1096_ng_umqv_Wholesale1.grib.923a8.idx'\n",
      "Traceback (most recent call last):\n",
      "  File \"/home/jack/miniconda3/envs/nowcasting_dataset/lib/python3.9/site-packages/cfgrib/messages.py\", line 239, in __iter__\n",
      "    yield self.message_from_file(file, errors=self.errors)\n",
      "  File \"/home/jack/miniconda3/envs/nowcasting_dataset/lib/python3.9/site-packages/cfgrib/messages.py\", line 255, in message_from_file\n",
      "    return self.message_class.from_file(file, offset, **kwargs)\n",
      "  File \"/home/jack/miniconda3/envs/nowcasting_dataset/lib/python3.9/site-packages/cfgrib/messages.py\", line 102, in from_file\n",
      "    raise EOFError(\"End of file: %r\" % file)\n",
      "EOFError: End of file: <_io.BufferedReader name='/media/jack/128GB_USB/NWPs/2019/06/15/201906150000_u1096_ng_umqv_Wholesale1.grib'>\n",
      "\n",
      "During handling of the above exception, another exception occurred:\n",
      "\n",
      "Traceback (most recent call last):\n",
      "  File \"/home/jack/miniconda3/envs/nowcasting_dataset/lib/python3.9/site-packages/cfgrib/messages.py\", line 349, in from_indexpath_or_filestream\n",
      "    self = cls.from_filestream(filestream, index_keys)\n",
      "  File \"/home/jack/miniconda3/envs/nowcasting_dataset/lib/python3.9/site-packages/cfgrib/messages.py\", line 298, in from_filestream\n",
      "    for message in filestream:\n",
      "  File \"/home/jack/miniconda3/envs/nowcasting_dataset/lib/python3.9/site-packages/cfgrib/messages.py\", line 243, in __iter__\n",
      "    raise EOFError(\"No valid message found in file: %r\" % self.path)\n",
      "EOFError: No valid message found in file: '/media/jack/128GB_USB/NWPs/2019/06/15/201906150000_u1096_ng_umqv_Wholesale1.grib'\n",
      "Can't read index file '/media/jack/128GB_USB/NWPs/2019/06/15/201906150000_u1096_ng_umqv_Wholesale1.grib.923a8.idx'\n",
      "Traceback (most recent call last):\n",
      "  File \"/home/jack/miniconda3/envs/nowcasting_dataset/lib/python3.9/site-packages/cfgrib/messages.py\", line 358, in from_indexpath_or_filestream\n",
      "    index_mtime = os.path.getmtime(indexpath)\n",
      "  File \"/home/jack/miniconda3/envs/nowcasting_dataset/lib/python3.9/genericpath.py\", line 55, in getmtime\n",
      "    return os.stat(filename).st_mtime\n",
      "FileNotFoundError: [Errno 2] No such file or directory: '/media/jack/128GB_USB/NWPs/2019/06/15/201906150000_u1096_ng_umqv_Wholesale1.grib.923a8.idx'\n",
      "Can't create file '/media/jack/128GB_USB/NWPs/2019/06/15/201906150300_u1096_ng_umqv_Wholesale1.grib.923a8.idx'\n",
      "Traceback (most recent call last):\n",
      "  File \"/home/jack/miniconda3/envs/nowcasting_dataset/lib/python3.9/site-packages/cfgrib/messages.py\", line 239, in __iter__\n",
      "    yield self.message_from_file(file, errors=self.errors)\n",
      "  File \"/home/jack/miniconda3/envs/nowcasting_dataset/lib/python3.9/site-packages/cfgrib/messages.py\", line 255, in message_from_file\n",
      "    return self.message_class.from_file(file, offset, **kwargs)\n",
      "  File \"/home/jack/miniconda3/envs/nowcasting_dataset/lib/python3.9/site-packages/cfgrib/messages.py\", line 102, in from_file\n",
      "    raise EOFError(\"End of file: %r\" % file)\n",
      "EOFError: End of file: <_io.BufferedReader name='/media/jack/128GB_USB/NWPs/2019/06/15/201906150300_u1096_ng_umqv_Wholesale1.grib'>\n",
      "\n",
      "During handling of the above exception, another exception occurred:\n",
      "\n",
      "Traceback (most recent call last):\n",
      "  File \"/home/jack/miniconda3/envs/nowcasting_dataset/lib/python3.9/site-packages/cfgrib/messages.py\", line 349, in from_indexpath_or_filestream\n",
      "    self = cls.from_filestream(filestream, index_keys)\n",
      "  File \"/home/jack/miniconda3/envs/nowcasting_dataset/lib/python3.9/site-packages/cfgrib/messages.py\", line 298, in from_filestream\n",
      "    for message in filestream:\n",
      "  File \"/home/jack/miniconda3/envs/nowcasting_dataset/lib/python3.9/site-packages/cfgrib/messages.py\", line 243, in __iter__\n",
      "    raise EOFError(\"No valid message found in file: %r\" % self.path)\n",
      "EOFError: No valid message found in file: '/media/jack/128GB_USB/NWPs/2019/06/15/201906150300_u1096_ng_umqv_Wholesale1.grib'\n",
      "Can't read index file '/media/jack/128GB_USB/NWPs/2019/06/15/201906150300_u1096_ng_umqv_Wholesale1.grib.923a8.idx'\n",
      "Traceback (most recent call last):\n",
      "  File \"/home/jack/miniconda3/envs/nowcasting_dataset/lib/python3.9/site-packages/cfgrib/messages.py\", line 358, in from_indexpath_or_filestream\n",
      "    index_mtime = os.path.getmtime(indexpath)\n",
      "  File \"/home/jack/miniconda3/envs/nowcasting_dataset/lib/python3.9/genericpath.py\", line 55, in getmtime\n",
      "    return os.stat(filename).st_mtime\n",
      "FileNotFoundError: [Errno 2] No such file or directory: '/media/jack/128GB_USB/NWPs/2019/06/15/201906150300_u1096_ng_umqv_Wholesale1.grib.923a8.idx'\n",
      "Can't create file '/media/jack/128GB_USB/NWPs/2019/06/15/201906150600_u1096_ng_umqv_Wholesale2.grib.923a8.idx'\n",
      "Traceback (most recent call last):\n",
      "  File \"/home/jack/miniconda3/envs/nowcasting_dataset/lib/python3.9/site-packages/cfgrib/messages.py\", line 239, in __iter__\n",
      "    yield self.message_from_file(file, errors=self.errors)\n",
      "  File \"/home/jack/miniconda3/envs/nowcasting_dataset/lib/python3.9/site-packages/cfgrib/messages.py\", line 255, in message_from_file\n",
      "    return self.message_class.from_file(file, offset, **kwargs)\n",
      "  File \"/home/jack/miniconda3/envs/nowcasting_dataset/lib/python3.9/site-packages/cfgrib/messages.py\", line 102, in from_file\n",
      "    raise EOFError(\"End of file: %r\" % file)\n",
      "EOFError: End of file: <_io.BufferedReader name='/media/jack/128GB_USB/NWPs/2019/06/15/201906150600_u1096_ng_umqv_Wholesale2.grib'>\n",
      "\n",
      "During handling of the above exception, another exception occurred:\n",
      "\n",
      "Traceback (most recent call last):\n",
      "  File \"/home/jack/miniconda3/envs/nowcasting_dataset/lib/python3.9/site-packages/cfgrib/messages.py\", line 349, in from_indexpath_or_filestream\n",
      "    self = cls.from_filestream(filestream, index_keys)\n",
      "  File \"/home/jack/miniconda3/envs/nowcasting_dataset/lib/python3.9/site-packages/cfgrib/messages.py\", line 298, in from_filestream\n",
      "    for message in filestream:\n",
      "  File \"/home/jack/miniconda3/envs/nowcasting_dataset/lib/python3.9/site-packages/cfgrib/messages.py\", line 243, in __iter__\n",
      "    raise EOFError(\"No valid message found in file: %r\" % self.path)\n",
      "EOFError: No valid message found in file: '/media/jack/128GB_USB/NWPs/2019/06/15/201906150600_u1096_ng_umqv_Wholesale2.grib'\n",
      "Can't read index file '/media/jack/128GB_USB/NWPs/2019/06/15/201906150600_u1096_ng_umqv_Wholesale2.grib.923a8.idx'\n",
      "Traceback (most recent call last):\n",
      "  File \"/home/jack/miniconda3/envs/nowcasting_dataset/lib/python3.9/site-packages/cfgrib/messages.py\", line 358, in from_indexpath_or_filestream\n",
      "    index_mtime = os.path.getmtime(indexpath)\n",
      "  File \"/home/jack/miniconda3/envs/nowcasting_dataset/lib/python3.9/genericpath.py\", line 55, in getmtime\n",
      "    return os.stat(filename).st_mtime\n",
      "FileNotFoundError: [Errno 2] No such file or directory: '/media/jack/128GB_USB/NWPs/2019/06/15/201906150600_u1096_ng_umqv_Wholesale2.grib.923a8.idx'\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Loading /media/jack/128GB_USB/NWPs/2018/12/31/201812312100_u1096_ng_umqv_Wholesale1.grib\n",
      "No valid message found in file: '/media/jack/128GB_USB/NWPs/2018/12/31/201812312100_u1096_ng_umqv_Wholesale1.grib' Filesize = 0 bytes\n",
      "Loading /media/jack/128GB_USB/NWPs/2019/06/15/201906150000_u1096_ng_umqv_Wholesale1.grib\n",
      "No valid message found in file: '/media/jack/128GB_USB/NWPs/2019/06/15/201906150000_u1096_ng_umqv_Wholesale1.grib' Filesize = 0 bytes\n",
      "Loading /media/jack/128GB_USB/NWPs/2019/06/15/201906150300_u1096_ng_umqv_Wholesale1.grib\n",
      "No valid message found in file: '/media/jack/128GB_USB/NWPs/2019/06/15/201906150300_u1096_ng_umqv_Wholesale1.grib' Filesize = 0 bytes\n",
      "Loading /media/jack/128GB_USB/NWPs/2019/06/15/201906150600_u1096_ng_umqv_Wholesale2.grib\n",
      "No valid message found in file: '/media/jack/128GB_USB/NWPs/2019/06/15/201906150600_u1096_ng_umqv_Wholesale2.grib' Filesize = 0 bytes\n",
      "Loading /media/jack/128GB_USB/NWPs/2019/06/15/201906150900_u1096_ng_umqv_Wholesale1.grib\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "ecCodes provides no latitudes/longitudes for gridType='transverse_mercator'\n",
      "ecCodes provides no latitudes/longitudes for gridType='transverse_mercator'\n",
      "ecCodes provides no latitudes/longitudes for gridType='transverse_mercator'\n",
      "ecCodes provides no latitudes/longitudes for gridType='transverse_mercator'\n",
      "ecCodes provides no latitudes/longitudes for gridType='transverse_mercator'\n",
      "ecCodes provides no latitudes/longitudes for gridType='transverse_mercator'\n",
      "ecCodes provides no latitudes/longitudes for gridType='transverse_mercator'\n",
      "ecCodes provides no latitudes/longitudes for gridType='transverse_mercator'\n",
      "ecCodes provides no latitudes/longitudes for gridType='transverse_mercator'\n",
      "ecCodes provides no latitudes/longitudes for gridType='transverse_mercator'\n",
      "Can't create file '/media/jack/128GB_USB/NWPs/2019/06/15/201906150900_u1096_ng_umqv_Wholesale2.grib.923a8.idx'\n",
      "Traceback (most recent call last):\n",
      "  File \"/home/jack/miniconda3/envs/nowcasting_dataset/lib/python3.9/site-packages/cfgrib/messages.py\", line 239, in __iter__\n",
      "    yield self.message_from_file(file, errors=self.errors)\n",
      "  File \"/home/jack/miniconda3/envs/nowcasting_dataset/lib/python3.9/site-packages/cfgrib/messages.py\", line 255, in message_from_file\n",
      "    return self.message_class.from_file(file, offset, **kwargs)\n",
      "  File \"/home/jack/miniconda3/envs/nowcasting_dataset/lib/python3.9/site-packages/cfgrib/messages.py\", line 102, in from_file\n",
      "    raise EOFError(\"End of file: %r\" % file)\n",
      "EOFError: End of file: <_io.BufferedReader name='/media/jack/128GB_USB/NWPs/2019/06/15/201906150900_u1096_ng_umqv_Wholesale2.grib'>\n",
      "\n",
      "During handling of the above exception, another exception occurred:\n",
      "\n",
      "Traceback (most recent call last):\n",
      "  File \"/home/jack/miniconda3/envs/nowcasting_dataset/lib/python3.9/site-packages/cfgrib/messages.py\", line 349, in from_indexpath_or_filestream\n",
      "    self = cls.from_filestream(filestream, index_keys)\n",
      "  File \"/home/jack/miniconda3/envs/nowcasting_dataset/lib/python3.9/site-packages/cfgrib/messages.py\", line 298, in from_filestream\n",
      "    for message in filestream:\n",
      "  File \"/home/jack/miniconda3/envs/nowcasting_dataset/lib/python3.9/site-packages/cfgrib/messages.py\", line 243, in __iter__\n",
      "    raise EOFError(\"No valid message found in file: %r\" % self.path)\n",
      "EOFError: No valid message found in file: '/media/jack/128GB_USB/NWPs/2019/06/15/201906150900_u1096_ng_umqv_Wholesale2.grib'\n",
      "Can't read index file '/media/jack/128GB_USB/NWPs/2019/06/15/201906150900_u1096_ng_umqv_Wholesale2.grib.923a8.idx'\n",
      "Traceback (most recent call last):\n",
      "  File \"/home/jack/miniconda3/envs/nowcasting_dataset/lib/python3.9/site-packages/cfgrib/messages.py\", line 358, in from_indexpath_or_filestream\n",
      "    index_mtime = os.path.getmtime(indexpath)\n",
      "  File \"/home/jack/miniconda3/envs/nowcasting_dataset/lib/python3.9/genericpath.py\", line 55, in getmtime\n",
      "    return os.stat(filename).st_mtime\n",
      "FileNotFoundError: [Errno 2] No such file or directory: '/media/jack/128GB_USB/NWPs/2019/06/15/201906150900_u1096_ng_umqv_Wholesale2.grib.923a8.idx'\n",
      "Can't create file '/media/jack/128GB_USB/NWPs/2019/06/15/201906151200_u1096_ng_umqv_Wholesale1.grib.923a8.idx'\n",
      "Traceback (most recent call last):\n",
      "  File \"/home/jack/miniconda3/envs/nowcasting_dataset/lib/python3.9/site-packages/cfgrib/messages.py\", line 239, in __iter__\n",
      "    yield self.message_from_file(file, errors=self.errors)\n",
      "  File \"/home/jack/miniconda3/envs/nowcasting_dataset/lib/python3.9/site-packages/cfgrib/messages.py\", line 255, in message_from_file\n",
      "    return self.message_class.from_file(file, offset, **kwargs)\n",
      "  File \"/home/jack/miniconda3/envs/nowcasting_dataset/lib/python3.9/site-packages/cfgrib/messages.py\", line 102, in from_file\n",
      "    raise EOFError(\"End of file: %r\" % file)\n",
      "EOFError: End of file: <_io.BufferedReader name='/media/jack/128GB_USB/NWPs/2019/06/15/201906151200_u1096_ng_umqv_Wholesale1.grib'>\n",
      "\n",
      "During handling of the above exception, another exception occurred:\n",
      "\n",
      "Traceback (most recent call last):\n",
      "  File \"/home/jack/miniconda3/envs/nowcasting_dataset/lib/python3.9/site-packages/cfgrib/messages.py\", line 349, in from_indexpath_or_filestream\n",
      "    self = cls.from_filestream(filestream, index_keys)\n",
      "  File \"/home/jack/miniconda3/envs/nowcasting_dataset/lib/python3.9/site-packages/cfgrib/messages.py\", line 298, in from_filestream\n",
      "    for message in filestream:\n",
      "  File \"/home/jack/miniconda3/envs/nowcasting_dataset/lib/python3.9/site-packages/cfgrib/messages.py\", line 243, in __iter__\n",
      "    raise EOFError(\"No valid message found in file: %r\" % self.path)\n",
      "EOFError: No valid message found in file: '/media/jack/128GB_USB/NWPs/2019/06/15/201906151200_u1096_ng_umqv_Wholesale1.grib'\n",
      "Can't read index file '/media/jack/128GB_USB/NWPs/2019/06/15/201906151200_u1096_ng_umqv_Wholesale1.grib.923a8.idx'\n",
      "Traceback (most recent call last):\n",
      "  File \"/home/jack/miniconda3/envs/nowcasting_dataset/lib/python3.9/site-packages/cfgrib/messages.py\", line 358, in from_indexpath_or_filestream\n",
      "    index_mtime = os.path.getmtime(indexpath)\n",
      "  File \"/home/jack/miniconda3/envs/nowcasting_dataset/lib/python3.9/genericpath.py\", line 55, in getmtime\n",
      "    return os.stat(filename).st_mtime\n",
      "FileNotFoundError: [Errno 2] No such file or directory: '/media/jack/128GB_USB/NWPs/2019/06/15/201906151200_u1096_ng_umqv_Wholesale1.grib.923a8.idx'\n",
      "Can't create file '/media/jack/128GB_USB/NWPs/2019/06/15/201906151500_u1096_ng_umqv_Wholesale1.grib.923a8.idx'\n",
      "Traceback (most recent call last):\n",
      "  File \"/home/jack/miniconda3/envs/nowcasting_dataset/lib/python3.9/site-packages/cfgrib/messages.py\", line 239, in __iter__\n",
      "    yield self.message_from_file(file, errors=self.errors)\n",
      "  File \"/home/jack/miniconda3/envs/nowcasting_dataset/lib/python3.9/site-packages/cfgrib/messages.py\", line 255, in message_from_file\n",
      "    return self.message_class.from_file(file, offset, **kwargs)\n",
      "  File \"/home/jack/miniconda3/envs/nowcasting_dataset/lib/python3.9/site-packages/cfgrib/messages.py\", line 102, in from_file\n",
      "    raise EOFError(\"End of file: %r\" % file)\n",
      "EOFError: End of file: <_io.BufferedReader name='/media/jack/128GB_USB/NWPs/2019/06/15/201906151500_u1096_ng_umqv_Wholesale1.grib'>\n",
      "\n",
      "During handling of the above exception, another exception occurred:\n",
      "\n",
      "Traceback (most recent call last):\n",
      "  File \"/home/jack/miniconda3/envs/nowcasting_dataset/lib/python3.9/site-packages/cfgrib/messages.py\", line 349, in from_indexpath_or_filestream\n",
      "    self = cls.from_filestream(filestream, index_keys)\n",
      "  File \"/home/jack/miniconda3/envs/nowcasting_dataset/lib/python3.9/site-packages/cfgrib/messages.py\", line 298, in from_filestream\n",
      "    for message in filestream:\n",
      "  File \"/home/jack/miniconda3/envs/nowcasting_dataset/lib/python3.9/site-packages/cfgrib/messages.py\", line 243, in __iter__\n",
      "    raise EOFError(\"No valid message found in file: %r\" % self.path)\n",
      "EOFError: No valid message found in file: '/media/jack/128GB_USB/NWPs/2019/06/15/201906151500_u1096_ng_umqv_Wholesale1.grib'\n",
      "Can't read index file '/media/jack/128GB_USB/NWPs/2019/06/15/201906151500_u1096_ng_umqv_Wholesale1.grib.923a8.idx'\n",
      "Traceback (most recent call last):\n",
      "  File \"/home/jack/miniconda3/envs/nowcasting_dataset/lib/python3.9/site-packages/cfgrib/messages.py\", line 358, in from_indexpath_or_filestream\n",
      "    index_mtime = os.path.getmtime(indexpath)\n",
      "  File \"/home/jack/miniconda3/envs/nowcasting_dataset/lib/python3.9/genericpath.py\", line 55, in getmtime\n",
      "    return os.stat(filename).st_mtime\n",
      "FileNotFoundError: [Errno 2] No such file or directory: '/media/jack/128GB_USB/NWPs/2019/06/15/201906151500_u1096_ng_umqv_Wholesale1.grib.923a8.idx'\n",
      "Can't create file '/media/jack/128GB_USB/NWPs/2019/06/15/201906151800_u1096_ng_umqv_Wholesale2.grib.923a8.idx'\n",
      "Traceback (most recent call last):\n",
      "  File \"/home/jack/miniconda3/envs/nowcasting_dataset/lib/python3.9/site-packages/cfgrib/messages.py\", line 239, in __iter__\n",
      "    yield self.message_from_file(file, errors=self.errors)\n",
      "  File \"/home/jack/miniconda3/envs/nowcasting_dataset/lib/python3.9/site-packages/cfgrib/messages.py\", line 255, in message_from_file\n",
      "    return self.message_class.from_file(file, offset, **kwargs)\n",
      "  File \"/home/jack/miniconda3/envs/nowcasting_dataset/lib/python3.9/site-packages/cfgrib/messages.py\", line 102, in from_file\n",
      "    raise EOFError(\"End of file: %r\" % file)\n",
      "EOFError: End of file: <_io.BufferedReader name='/media/jack/128GB_USB/NWPs/2019/06/15/201906151800_u1096_ng_umqv_Wholesale2.grib'>\n",
      "\n",
      "During handling of the above exception, another exception occurred:\n",
      "\n",
      "Traceback (most recent call last):\n",
      "  File \"/home/jack/miniconda3/envs/nowcasting_dataset/lib/python3.9/site-packages/cfgrib/messages.py\", line 349, in from_indexpath_or_filestream\n",
      "    self = cls.from_filestream(filestream, index_keys)\n",
      "  File \"/home/jack/miniconda3/envs/nowcasting_dataset/lib/python3.9/site-packages/cfgrib/messages.py\", line 298, in from_filestream\n",
      "    for message in filestream:\n",
      "  File \"/home/jack/miniconda3/envs/nowcasting_dataset/lib/python3.9/site-packages/cfgrib/messages.py\", line 243, in __iter__\n",
      "    raise EOFError(\"No valid message found in file: %r\" % self.path)\n",
      "EOFError: No valid message found in file: '/media/jack/128GB_USB/NWPs/2019/06/15/201906151800_u1096_ng_umqv_Wholesale2.grib'\n",
      "Can't read index file '/media/jack/128GB_USB/NWPs/2019/06/15/201906151800_u1096_ng_umqv_Wholesale2.grib.923a8.idx'\n",
      "Traceback (most recent call last):\n",
      "  File \"/home/jack/miniconda3/envs/nowcasting_dataset/lib/python3.9/site-packages/cfgrib/messages.py\", line 358, in from_indexpath_or_filestream\n",
      "    index_mtime = os.path.getmtime(indexpath)\n",
      "  File \"/home/jack/miniconda3/envs/nowcasting_dataset/lib/python3.9/genericpath.py\", line 55, in getmtime\n",
      "    return os.stat(filename).st_mtime\n",
      "FileNotFoundError: [Errno 2] No such file or directory: '/media/jack/128GB_USB/NWPs/2019/06/15/201906151800_u1096_ng_umqv_Wholesale2.grib.923a8.idx'\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Loading /media/jack/128GB_USB/NWPs/2019/06/15/201906150900_u1096_ng_umqv_Wholesale2.grib\n",
      "No valid message found in file: '/media/jack/128GB_USB/NWPs/2019/06/15/201906150900_u1096_ng_umqv_Wholesale2.grib' Filesize = 0 bytes\n",
      "Loading /media/jack/128GB_USB/NWPs/2019/06/15/201906151200_u1096_ng_umqv_Wholesale1.grib\n",
      "No valid message found in file: '/media/jack/128GB_USB/NWPs/2019/06/15/201906151200_u1096_ng_umqv_Wholesale1.grib' Filesize = 0 bytes\n",
      "Loading /media/jack/128GB_USB/NWPs/2019/06/15/201906151500_u1096_ng_umqv_Wholesale1.grib\n",
      "No valid message found in file: '/media/jack/128GB_USB/NWPs/2019/06/15/201906151500_u1096_ng_umqv_Wholesale1.grib' Filesize = 0 bytes\n",
      "Loading /media/jack/128GB_USB/NWPs/2019/06/15/201906151800_u1096_ng_umqv_Wholesale2.grib\n",
      "No valid message found in file: '/media/jack/128GB_USB/NWPs/2019/06/15/201906151800_u1096_ng_umqv_Wholesale2.grib' Filesize = 0 bytes\n",
      "Loading /media/jack/128GB_USB/NWPs/2019/06/15/201906152100_u1096_ng_umqv_Wholesale1.grib\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "ecCodes provides no latitudes/longitudes for gridType='transverse_mercator'\n",
      "ecCodes provides no latitudes/longitudes for gridType='transverse_mercator'\n",
      "ecCodes provides no latitudes/longitudes for gridType='transverse_mercator'\n",
      "ecCodes provides no latitudes/longitudes for gridType='transverse_mercator'\n",
      "ecCodes provides no latitudes/longitudes for gridType='transverse_mercator'\n",
      "ecCodes provides no latitudes/longitudes for gridType='transverse_mercator'\n",
      "ecCodes provides no latitudes/longitudes for gridType='transverse_mercator'\n",
      "ecCodes provides no latitudes/longitudes for gridType='transverse_mercator'\n",
      "ecCodes provides no latitudes/longitudes for gridType='transverse_mercator'\n",
      "ecCodes provides no latitudes/longitudes for gridType='transverse_mercator'\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Loading /media/jack/128GB_USB/NWPs/2019/06/15/201906152100_u1096_ng_umqv_Wholesale2.grib\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "ecCodes provides no latitudes/longitudes for gridType='transverse_mercator'\n",
      "ecCodes provides no latitudes/longitudes for gridType='transverse_mercator'\n",
      "ecCodes provides no latitudes/longitudes for gridType='transverse_mercator'\n",
      "ecCodes provides no latitudes/longitudes for gridType='transverse_mercator'\n",
      "ecCodes provides no latitudes/longitudes for gridType='transverse_mercator'\n",
      "ecCodes provides no latitudes/longitudes for gridType='transverse_mercator'\n",
      "ecCodes provides no latitudes/longitudes for gridType='transverse_mercator'\n",
      "ecCodes provides no latitudes/longitudes for gridType='transverse_mercator'\n",
      "ecCodes provides no latitudes/longitudes for gridType='transverse_mercator'\n",
      "ecCodes provides no latitudes/longitudes for gridType='transverse_mercator'\n",
      "ecCodes provides no latitudes/longitudes for gridType='transverse_mercator'\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Merging...\n",
      "Reshaping...\n",
      "Saving to Zarr...\n",
      "Writing to disk...\n",
      "Loading /media/jack/128GB_USB/NWPs/2021/09/30/202109300000_u1096_ng_umqv_Wholesale2.grib\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "ecCodes provides no latitudes/longitudes for gridType='transverse_mercator'\n",
      "ecCodes provides no latitudes/longitudes for gridType='transverse_mercator'\n",
      "ecCodes provides no latitudes/longitudes for gridType='transverse_mercator'\n",
      "ecCodes provides no latitudes/longitudes for gridType='transverse_mercator'\n",
      "ecCodes provides no latitudes/longitudes for gridType='transverse_mercator'\n",
      "ecCodes provides no latitudes/longitudes for gridType='transverse_mercator'\n",
      "ecCodes provides no latitudes/longitudes for gridType='transverse_mercator'\n",
      "ecCodes provides no latitudes/longitudes for gridType='transverse_mercator'\n",
      "ecCodes provides no latitudes/longitudes for gridType='transverse_mercator'\n",
      "ecCodes provides no latitudes/longitudes for gridType='transverse_mercator'\n",
      "ecCodes provides no latitudes/longitudes for gridType='transverse_mercator'\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Loading /media/jack/128GB_USB/NWPs/2021/09/30/202109300000_u1096_ng_umqv_Wholesale1.grib\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "ecCodes provides no latitudes/longitudes for gridType='transverse_mercator'\n",
      "ecCodes provides no latitudes/longitudes for gridType='transverse_mercator'\n",
      "ecCodes provides no latitudes/longitudes for gridType='transverse_mercator'\n",
      "ecCodes provides no latitudes/longitudes for gridType='transverse_mercator'\n",
      "ecCodes provides no latitudes/longitudes for gridType='transverse_mercator'\n",
      "ecCodes provides no latitudes/longitudes for gridType='transverse_mercator'\n",
      "ecCodes provides no latitudes/longitudes for gridType='transverse_mercator'\n",
      "ecCodes provides no latitudes/longitudes for gridType='transverse_mercator'\n",
      "ecCodes provides no latitudes/longitudes for gridType='transverse_mercator'\n",
      "ecCodes provides no latitudes/longitudes for gridType='transverse_mercator'\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Merging...\n",
      "Reshaping...\n",
      "Saving to Zarr...\n",
      "Writing to disk...\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Can't create file '/media/jack/128GB_USB/NWPs/2021/09/30/202109300300_u1096_ng_umqv_Wholesale2.grib.923a8.idx'\n",
      "Traceback (most recent call last):\n",
      "  File \"/home/jack/miniconda3/envs/nowcasting_dataset/lib/python3.9/site-packages/cfgrib/messages.py\", line 239, in __iter__\n",
      "    yield self.message_from_file(file, errors=self.errors)\n",
      "  File \"/home/jack/miniconda3/envs/nowcasting_dataset/lib/python3.9/site-packages/cfgrib/messages.py\", line 255, in message_from_file\n",
      "    return self.message_class.from_file(file, offset, **kwargs)\n",
      "  File \"/home/jack/miniconda3/envs/nowcasting_dataset/lib/python3.9/site-packages/cfgrib/messages.py\", line 102, in from_file\n",
      "    raise EOFError(\"End of file: %r\" % file)\n",
      "EOFError: End of file: <_io.BufferedReader name='/media/jack/128GB_USB/NWPs/2021/09/30/202109300300_u1096_ng_umqv_Wholesale2.grib'>\n",
      "\n",
      "During handling of the above exception, another exception occurred:\n",
      "\n",
      "Traceback (most recent call last):\n",
      "  File \"/home/jack/miniconda3/envs/nowcasting_dataset/lib/python3.9/site-packages/cfgrib/messages.py\", line 349, in from_indexpath_or_filestream\n",
      "    self = cls.from_filestream(filestream, index_keys)\n",
      "  File \"/home/jack/miniconda3/envs/nowcasting_dataset/lib/python3.9/site-packages/cfgrib/messages.py\", line 298, in from_filestream\n",
      "    for message in filestream:\n",
      "  File \"/home/jack/miniconda3/envs/nowcasting_dataset/lib/python3.9/site-packages/cfgrib/messages.py\", line 243, in __iter__\n",
      "    raise EOFError(\"No valid message found in file: %r\" % self.path)\n",
      "EOFError: No valid message found in file: '/media/jack/128GB_USB/NWPs/2021/09/30/202109300300_u1096_ng_umqv_Wholesale2.grib'\n",
      "Can't read index file '/media/jack/128GB_USB/NWPs/2021/09/30/202109300300_u1096_ng_umqv_Wholesale2.grib.923a8.idx'\n",
      "Traceback (most recent call last):\n",
      "  File \"/home/jack/miniconda3/envs/nowcasting_dataset/lib/python3.9/site-packages/cfgrib/messages.py\", line 358, in from_indexpath_or_filestream\n",
      "    index_mtime = os.path.getmtime(indexpath)\n",
      "  File \"/home/jack/miniconda3/envs/nowcasting_dataset/lib/python3.9/genericpath.py\", line 55, in getmtime\n",
      "    return os.stat(filename).st_mtime\n",
      "FileNotFoundError: [Errno 2] No such file or directory: '/media/jack/128GB_USB/NWPs/2021/09/30/202109300300_u1096_ng_umqv_Wholesale2.grib.923a8.idx'\n",
      "Can't create file '/media/jack/128GB_USB/NWPs/2021/09/30/202109300600_u1096_ng_umqv_Wholesale1.grib.923a8.idx'\n",
      "Traceback (most recent call last):\n",
      "  File \"/home/jack/miniconda3/envs/nowcasting_dataset/lib/python3.9/site-packages/cfgrib/messages.py\", line 239, in __iter__\n",
      "    yield self.message_from_file(file, errors=self.errors)\n",
      "  File \"/home/jack/miniconda3/envs/nowcasting_dataset/lib/python3.9/site-packages/cfgrib/messages.py\", line 255, in message_from_file\n",
      "    return self.message_class.from_file(file, offset, **kwargs)\n",
      "  File \"/home/jack/miniconda3/envs/nowcasting_dataset/lib/python3.9/site-packages/cfgrib/messages.py\", line 102, in from_file\n",
      "    raise EOFError(\"End of file: %r\" % file)\n",
      "EOFError: End of file: <_io.BufferedReader name='/media/jack/128GB_USB/NWPs/2021/09/30/202109300600_u1096_ng_umqv_Wholesale1.grib'>\n",
      "\n",
      "During handling of the above exception, another exception occurred:\n",
      "\n",
      "Traceback (most recent call last):\n",
      "  File \"/home/jack/miniconda3/envs/nowcasting_dataset/lib/python3.9/site-packages/cfgrib/messages.py\", line 349, in from_indexpath_or_filestream\n",
      "    self = cls.from_filestream(filestream, index_keys)\n",
      "  File \"/home/jack/miniconda3/envs/nowcasting_dataset/lib/python3.9/site-packages/cfgrib/messages.py\", line 298, in from_filestream\n",
      "    for message in filestream:\n",
      "  File \"/home/jack/miniconda3/envs/nowcasting_dataset/lib/python3.9/site-packages/cfgrib/messages.py\", line 243, in __iter__\n",
      "    raise EOFError(\"No valid message found in file: %r\" % self.path)\n",
      "EOFError: No valid message found in file: '/media/jack/128GB_USB/NWPs/2021/09/30/202109300600_u1096_ng_umqv_Wholesale1.grib'\n",
      "Can't read index file '/media/jack/128GB_USB/NWPs/2021/09/30/202109300600_u1096_ng_umqv_Wholesale1.grib.923a8.idx'\n",
      "Traceback (most recent call last):\n",
      "  File \"/home/jack/miniconda3/envs/nowcasting_dataset/lib/python3.9/site-packages/cfgrib/messages.py\", line 358, in from_indexpath_or_filestream\n",
      "    index_mtime = os.path.getmtime(indexpath)\n",
      "  File \"/home/jack/miniconda3/envs/nowcasting_dataset/lib/python3.9/genericpath.py\", line 55, in getmtime\n",
      "    return os.stat(filename).st_mtime\n",
      "FileNotFoundError: [Errno 2] No such file or directory: '/media/jack/128GB_USB/NWPs/2021/09/30/202109300600_u1096_ng_umqv_Wholesale1.grib.923a8.idx'\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Loading /media/jack/128GB_USB/NWPs/2021/09/30/202109300300_u1096_ng_umqv_Wholesale2.grib\n",
      "No valid message found in file: '/media/jack/128GB_USB/NWPs/2021/09/30/202109300300_u1096_ng_umqv_Wholesale2.grib' Filesize = 0 bytes\n",
      "Loading /media/jack/128GB_USB/NWPs/2021/09/30/202109300600_u1096_ng_umqv_Wholesale1.grib\n",
      "No valid message found in file: '/media/jack/128GB_USB/NWPs/2021/09/30/202109300600_u1096_ng_umqv_Wholesale1.grib' Filesize = 0 bytes\n",
      "Done!\n"
     ]
    }
   ],
   "source": [
    "for nwp_init_datetime_utc in map_datetime_to_grib_filename.index.unique():\n",
    "    # Sanity check!\n",
    "    indicies = np.where(map_datetime_to_grib_filename.index == nwp_init_datetime_utc)[0]\n",
    "    n_filenames = len(indicies)\n",
    "    if n_filenames != 2:\n",
    "        print(n_filenames, \"filenames found for\", nwp_init_datetime_utc, \".  Expected 2. Skipping!\")\n",
    "        continue\n",
    "        \n",
    "    # Load files.\n",
    "    full_filenames = map_datetime_to_grib_filename[nwp_init_datetime_utc]\n",
    "    ds = load_grib_for_single_nwp_init_time(full_filenames)\n",
    "    \n",
    "    # Save to Zarr.\n",
    "    if ds is not None:\n",
    "        print(\"Saving to Zarr...\")\n",
    "        append_to_zarr(\n",
    "            ds, \n",
    "            #  \"/home/jack/data/nwp.zarr\"\n",
    "            \"/mnt/storage_b/data/ocf/solar_pv_nowcasting/nowcasting_dataset_pipeline/NWP/UK_Met_Office/UKV/zarr/test.zarr\",\n",
    "        )\n",
    "        \n",
    "print(\"Done!\")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "b366603a-3421-4621-a6bb-33910956d893",
   "metadata": {},
   "source": [
    "## Load from Zarr"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "633fb9bd-da4b-4e06-acec-bc09bd41b214",
   "metadata": {},
   "outputs": [],
   "source": [
    "ds_from_zarr = xr.open_dataset(\"/home/jack/data/nwp.zarr\", engine=\"zarr\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "82e01ee2-c24e-44ed-a085-67cc35fbc06f",
   "metadata": {},
   "outputs": [],
   "source": [
    "ds_from_zarr"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "3d48a95b-7b89-4ce4-8151-e5aef7a840fc",
   "metadata": {},
   "outputs": [],
   "source": [
    "ds_from_zarr['t'].isel(step=6, time=1).plot.imshow()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "f00995ad-5c1f-41fe-87ba-7ab678258add",
   "metadata": {},
   "outputs": [],
   "source": [
    "ds_from_zarr['t'].encoding"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "1e172c71-0222-47ab-a46d-f58078649962",
   "metadata": {},
   "outputs": [],
   "source": [
    "ds_from_zarr['t']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "cab8ba76-7aa3-4431-be48-1a561f56e3d4",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "nowcasting_dataset",
   "language": "python",
   "name": "nowcasting_dataset"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.7"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
